{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data science works (Project RNA editing)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential editing analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import os\n",
    "\n",
    "# vars\n",
    "WORKDIR = \"/home/b05b01002/HDD3/project_RE/\"\n",
    "PARAMS = \"configs/differential_editing/Egr.json\"\n",
    "\n",
    "# change work direction\n",
    "os.chdir(WORKDIR)\n",
    "params = json.load(open(PARAMS))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import pvalue.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>xylem:phloem</th>\n",
       "      <th>xylem:leaf</th>\n",
       "      <th>xylem:shoot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>100092</td>\n",
       "      <td>TG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>100093</td>\n",
       "      <td>TG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10412169</td>\n",
       "      <td>TC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10413299</td>\n",
       "      <td>TC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10414032</td>\n",
       "      <td>AG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190215</th>\n",
       "      <td>scaffold_987</td>\n",
       "      <td>8590</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.830</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190216</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>2775</td>\n",
       "      <td>GA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190217</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1722</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190218</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1376</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.990</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190219</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1693</td>\n",
       "      <td>AC</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190220 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Region  Position Subtype  xylem:phloem  xylem:leaf  xylem:shoot\n",
       "0              Chr01    100092      TG           NaN         NaN          NaN\n",
       "1              Chr01    100093      TG           NaN         NaN          NaN\n",
       "2              Chr01  10412169      TC           NaN         NaN          NaN\n",
       "3              Chr01  10413299      TC           NaN         NaN          NaN\n",
       "4              Chr01  10414032      AG           NaN         NaN          NaN\n",
       "...              ...       ...     ...           ...         ...          ...\n",
       "190215  scaffold_987      8590      AG         0.830         NaN          NaN\n",
       "190216  scaffold_991      2775      GA           NaN         NaN          NaN\n",
       "190217  scaffold_991      1722      AG         0.460       0.576        0.901\n",
       "190218  scaffold_991      1376      CA         0.100       0.990        1.000\n",
       "190219  scaffold_991      1693      AC         0.522       0.491        0.901\n",
       "\n",
       "[190220 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pvalues = pd.read_csv(\n",
    "    params[\"outdir\"],\n",
    "    names = [\n",
    "        \"Region\",\n",
    "        \"Position\",\n",
    "        \"Subtype\"\n",
    "    ] + params[\"pairing\"]\n",
    ")\n",
    "pvalues"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exclude rows with 3 NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>xylem:phloem</th>\n",
       "      <th>xylem:leaf</th>\n",
       "      <th>xylem:shoot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10413507</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10482731</td>\n",
       "      <td>AC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10482834</td>\n",
       "      <td>CA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10482931</td>\n",
       "      <td>TC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10483476</td>\n",
       "      <td>GT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.429</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190214</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>13333</td>\n",
       "      <td>AG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.606</td>\n",
       "      <td>0.176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190215</th>\n",
       "      <td>scaffold_987</td>\n",
       "      <td>8590</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.830</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190217</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1722</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190218</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1376</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.990</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190219</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1693</td>\n",
       "      <td>AC</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133579 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Region  Position Subtype  xylem:phloem  xylem:leaf  xylem:shoot\n",
       "7              Chr01  10413507      AG         1.000       0.000        0.126\n",
       "8              Chr01  10482731      AC           NaN       0.070        0.112\n",
       "9              Chr01  10482834      CA           NaN       0.095        0.121\n",
       "10             Chr01  10482931      TC           NaN       1.000        0.183\n",
       "11             Chr01  10483476      GT           NaN       0.429        1.000\n",
       "...              ...       ...     ...           ...         ...          ...\n",
       "190214  scaffold_991     13333      AG           NaN       0.606        0.176\n",
       "190215  scaffold_987      8590      AG         0.830         NaN          NaN\n",
       "190217  scaffold_991      1722      AG         0.460       0.576        0.901\n",
       "190218  scaffold_991      1376      CA         0.100       0.990        1.000\n",
       "190219  scaffold_991      1693      AC         0.522       0.491        0.901\n",
       "\n",
       "[133579 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "na_3dropped = pvalues.dropna(axis=0, thresh=1, subset=params[\"pairing\"])\n",
    "na_3dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exclude rows with NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>xylem:phloem</th>\n",
       "      <th>xylem:leaf</th>\n",
       "      <th>xylem:shoot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10413507</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10490455</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.149</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10413587</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.662</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10491752</td>\n",
       "      <td>TC</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.101</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>10473468</td>\n",
       "      <td>GA</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190212</th>\n",
       "      <td>scaffold_987</td>\n",
       "      <td>7682</td>\n",
       "      <td>TC</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190213</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1245</td>\n",
       "      <td>CG</td>\n",
       "      <td>0.896</td>\n",
       "      <td>0.475</td>\n",
       "      <td>0.592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190217</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1722</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190218</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1376</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.990</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190219</th>\n",
       "      <td>scaffold_991</td>\n",
       "      <td>1693</td>\n",
       "      <td>AC</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.491</td>\n",
       "      <td>0.901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98787 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Region  Position Subtype  xylem:phloem  xylem:leaf  xylem:shoot\n",
       "7              Chr01  10413507      AG         1.000       0.000        0.126\n",
       "13             Chr01  10490455      TC         0.149       1.000        1.000\n",
       "14             Chr01  10413587      TC         0.662       0.000        0.237\n",
       "15             Chr01  10491752      TC         1.000       0.101        1.000\n",
       "16             Chr01  10473468      GA         1.000       1.000        0.171\n",
       "...              ...       ...     ...           ...         ...          ...\n",
       "190212  scaffold_987      7682      TC         1.000       0.998        0.121\n",
       "190213  scaffold_991      1245      CG         0.896       0.475        0.592\n",
       "190217  scaffold_991      1722      AG         0.460       0.576        0.901\n",
       "190218  scaffold_991      1376      CA         0.100       0.990        1.000\n",
       "190219  scaffold_991      1693      AC         0.522       0.491        0.901\n",
       "\n",
       "[98787 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "na_dropped = pvalues.dropna(axis=0)\n",
    "na_dropped"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "find positions where all p-value < 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>xylem:phloem</th>\n",
       "      <th>xylem:leaf</th>\n",
       "      <th>xylem:shoot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>11116387</td>\n",
       "      <td>CT</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>11117874</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12173000</td>\n",
       "      <td>CT</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12208318</td>\n",
       "      <td>AC</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12208319</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189536</th>\n",
       "      <td>scaffold_587</td>\n",
       "      <td>1169</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189558</th>\n",
       "      <td>scaffold_587</td>\n",
       "      <td>38562</td>\n",
       "      <td>CA</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189613</th>\n",
       "      <td>scaffold_621</td>\n",
       "      <td>12154</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189803</th>\n",
       "      <td>scaffold_70</td>\n",
       "      <td>70953</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190139</th>\n",
       "      <td>scaffold_942</td>\n",
       "      <td>7540</td>\n",
       "      <td>CT</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2420 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Region  Position Subtype  xylem:phloem  xylem:leaf  xylem:shoot\n",
       "87             Chr01  11116387      CT         0.005       0.001        0.003\n",
       "95             Chr01  11117874      AG         0.040       0.005        0.030\n",
       "463            Chr01  12173000      CT         0.011       0.015        0.016\n",
       "495            Chr01  12208318      AC         0.010       0.031        0.015\n",
       "512            Chr01  12208319      TC         0.014       0.039        0.020\n",
       "...              ...       ...     ...           ...         ...          ...\n",
       "189536  scaffold_587      1169      TC         0.025       0.003        0.004\n",
       "189558  scaffold_587     38562      CA         0.004       0.024        0.019\n",
       "189613  scaffold_621     12154      TC         0.011       0.015        0.004\n",
       "189803   scaffold_70     70953      GA         0.013       0.041        0.002\n",
       "190139  scaffold_942      7540      CT         0.038       0.000        0.001\n",
       "\n",
       "[2420 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xylem_spec = na_dropped[\n",
    "    (na_dropped[\"xylem:phloem\"] < 0.05) &\n",
    "    (na_dropped[\"xylem:leaf\"] < 0.05) &\n",
    "    (na_dropped[\"xylem:shoot\"] < 0.05)\n",
    "]\n",
    "xylem_spec.sort_values(by=[\"Region\", \"Position\"])\n",
    "xylem_spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "xylem_spec.to_csv(\"temp/egr_xylem_specific.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotated xylem-specific editing sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Chr01</th>\n",
       "      <th>100092</th>\n",
       "      <td>100092</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>Eucgr.A00216.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100093</th>\n",
       "      <td>100093</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>Eucgr.A00216.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10412169</th>\n",
       "      <td>10412169</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00959.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon6:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10413299</th>\n",
       "      <td>10413299</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00959.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon9:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10414032</th>\n",
       "      <td>10414032</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00959.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon11:c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_987</th>\n",
       "      <th>8590</th>\n",
       "      <td>8590</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;NONE</td>\n",
       "      <td>dist=NONE;dist=NONE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">scaffold_991</th>\n",
       "      <th>2775</th>\n",
       "      <td>2775</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Eucgr.L02525.v2.0;NONE</td>\n",
       "      <td>dist=1411;dist=NONE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1722</th>\n",
       "      <td>1722</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream</td>\n",
       "      <td>Eucgr.L02525.v2.0</td>\n",
       "      <td>dist=358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>1376</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream</td>\n",
       "      <td>Eucgr.L02525.v2.0</td>\n",
       "      <td>dist=12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1693</th>\n",
       "      <td>1693</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>upstream</td>\n",
       "      <td>Eucgr.L02525.v2.0</td>\n",
       "      <td>dist=329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190220 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            End Ref Alt Func.refGene            Gene.refGene  \\\n",
       "Chr          Start                                                             \n",
       "Chr01        100092      100092   T   G     intronic       Eucgr.A00216.v2.0   \n",
       "             100093      100093   T   G     intronic       Eucgr.A00216.v2.0   \n",
       "             10412169  10412169   T   C       exonic       Eucgr.A00959.v2.0   \n",
       "             10413299  10413299   T   C       exonic       Eucgr.A00959.v2.0   \n",
       "             10414032  10414032   A   G       exonic       Eucgr.A00959.v2.0   \n",
       "...                         ...  ..  ..          ...                     ...   \n",
       "scaffold_987 8590          8590   A   G   intergenic               NONE;NONE   \n",
       "scaffold_991 2775          2775   G   A   intergenic  Eucgr.L02525.v2.0;NONE   \n",
       "             1722          1722   A   G     upstream       Eucgr.L02525.v2.0   \n",
       "             1376          1376   C   A     upstream       Eucgr.L02525.v2.0   \n",
       "             1693          1693   A   C     upstream       Eucgr.L02525.v2.0   \n",
       "\n",
       "                        GeneDetail.refGene ExonicFunc.refGene  \\\n",
       "Chr          Start                                              \n",
       "Chr01        100092                    NaN                NaN   \n",
       "             100093                    NaN                NaN   \n",
       "             10412169                  NaN  nonsynonymous SNV   \n",
       "             10413299                  NaN     synonymous SNV   \n",
       "             10414032                  NaN     synonymous SNV   \n",
       "...                                    ...                ...   \n",
       "scaffold_987 8590      dist=NONE;dist=NONE                NaN   \n",
       "scaffold_991 2775      dist=1411;dist=NONE                NaN   \n",
       "             1722                 dist=358                NaN   \n",
       "             1376                  dist=12                NaN   \n",
       "             1693                 dist=329                NaN   \n",
       "\n",
       "                                                        AAChange.refGene  \n",
       "Chr          Start                                                        \n",
       "Chr01        100092                                                  NaN  \n",
       "             100093                                                  NaN  \n",
       "             10412169  Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon6:c....  \n",
       "             10413299  Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon9:c....  \n",
       "             10414032  Eucgr.A00959.v2.0:Eucgr.A00959.1.v2.0:exon11:c...  \n",
       "...                                                                  ...  \n",
       "scaffold_987 8590                                                    NaN  \n",
       "scaffold_991 2775                                                    NaN  \n",
       "             1722                                                    NaN  \n",
       "             1376                                                    NaN  \n",
       "             1693                                                    NaN  \n",
       "\n",
       "[190220 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated = pd.read_csv(\n",
    "    \"/home/b05b01002/HDD3/project_RE/test/ANNOVAR-test/final.Egr_multianno.txt\", \n",
    "    sep = \"\\t\"\n",
    ")\n",
    "annotated.set_index([\"Chr\", \"Start\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Subtype</th>\n",
       "      <th>xylem:phloem</th>\n",
       "      <th>xylem:leaf</th>\n",
       "      <th>xylem:shoot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Position</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Chr01</th>\n",
       "      <th>11116387</th>\n",
       "      <td>CT</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11117874</th>\n",
       "      <td>AG</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12173000</th>\n",
       "      <td>CT</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12208318</th>\n",
       "      <td>AC</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12208319</th>\n",
       "      <td>TC</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">scaffold_587</th>\n",
       "      <th>1169</th>\n",
       "      <td>TC</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38562</th>\n",
       "      <td>CA</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_621</th>\n",
       "      <th>12154</th>\n",
       "      <td>TC</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_70</th>\n",
       "      <th>70953</th>\n",
       "      <td>GA</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_942</th>\n",
       "      <th>7540</th>\n",
       "      <td>CT</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2420 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Subtype  xylem:phloem  xylem:leaf  xylem:shoot\n",
       "Region       Position                                               \n",
       "Chr01        11116387      CT         0.005       0.001        0.003\n",
       "             11117874      AG         0.040       0.005        0.030\n",
       "             12173000      CT         0.011       0.015        0.016\n",
       "             12208318      AC         0.010       0.031        0.015\n",
       "             12208319      TC         0.014       0.039        0.020\n",
       "...                       ...           ...         ...          ...\n",
       "scaffold_587 1169          TC         0.025       0.003        0.004\n",
       "             38562         CA         0.004       0.024        0.019\n",
       "scaffold_621 12154         TC         0.011       0.015        0.004\n",
       "scaffold_70  70953         GA         0.013       0.041        0.002\n",
       "scaffold_942 7540          CT         0.038       0.000        0.001\n",
       "\n",
       "[2420 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xylem_spec.set_index([\"Region\", \"Position\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>11116387</td>\n",
       "      <td>11116387</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>Eucgr.A00907.v2.0</td>\n",
       "      <td>Eucgr.A00907.1.v2.0:c.-17C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>11117874</td>\n",
       "      <td>11117874</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>Eucgr.A00907.v2.0</td>\n",
       "      <td>Eucgr.A00907.1.v2.0:c.*19A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12173000</td>\n",
       "      <td>12173000</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00810.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>Eucgr.A00810.v2.0:Eucgr.A00810.1.v2.0:exon1:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12208318</td>\n",
       "      <td>12208318</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00807.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>Eucgr.A00807.v2.0:Eucgr.A00807.1.v2.0:exon1:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>12208319</td>\n",
       "      <td>12208319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.A00807.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>Eucgr.A00807.v2.0:Eucgr.A00807.1.v2.0:exon1:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189536</th>\n",
       "      <td>scaffold_587</td>\n",
       "      <td>1169</td>\n",
       "      <td>1169</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;Eucgr.L02107.v2.0</td>\n",
       "      <td>dist=NONE;dist=34988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189558</th>\n",
       "      <td>scaffold_587</td>\n",
       "      <td>38562</td>\n",
       "      <td>38562</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.L02107.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>Eucgr.L02107.v2.0:Eucgr.L02107.1.v2.0:exon1:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189613</th>\n",
       "      <td>scaffold_621</td>\n",
       "      <td>12154</td>\n",
       "      <td>12154</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.L02145.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>Eucgr.L02145.v2.0:Eucgr.L02145.1.v2.0:exon2:c....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189803</th>\n",
       "      <td>scaffold_70</td>\n",
       "      <td>70953</td>\n",
       "      <td>70953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>Eucgr.L00910.v2.0;Eucgr.L00913.v2.0</td>\n",
       "      <td>dist=38297;dist=23593</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190139</th>\n",
       "      <td>scaffold_942</td>\n",
       "      <td>7540</td>\n",
       "      <td>7540</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>Eucgr.L02476.v2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>Eucgr.L02476.v2.0:Eucgr.L02476.1.v2.0:exon5:c....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2420 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chr     Start       End Ref Alt Func.refGene  \\\n",
       "87             Chr01  11116387  11116387   C   T         UTR5   \n",
       "95             Chr01  11117874  11117874   A   G         UTR3   \n",
       "463            Chr01  12173000  12173000   C   T       exonic   \n",
       "495            Chr01  12208318  12208318   A   C       exonic   \n",
       "512            Chr01  12208319  12208319   T   C       exonic   \n",
       "...              ...       ...       ...  ..  ..          ...   \n",
       "189536  scaffold_587      1169      1169   T   C   intergenic   \n",
       "189558  scaffold_587     38562     38562   C   A       exonic   \n",
       "189613  scaffold_621     12154     12154   T   C       exonic   \n",
       "189803   scaffold_70     70953     70953   G   A   intergenic   \n",
       "190139  scaffold_942      7540      7540   C   T       exonic   \n",
       "\n",
       "                               Gene.refGene            GeneDetail.refGene  \\\n",
       "87                        Eucgr.A00907.v2.0  Eucgr.A00907.1.v2.0:c.-17C>T   \n",
       "95                        Eucgr.A00907.v2.0  Eucgr.A00907.1.v2.0:c.*19A>G   \n",
       "463                       Eucgr.A00810.v2.0                           NaN   \n",
       "495                       Eucgr.A00807.v2.0                           NaN   \n",
       "512                       Eucgr.A00807.v2.0                           NaN   \n",
       "...                                     ...                           ...   \n",
       "189536               NONE;Eucgr.L02107.v2.0          dist=NONE;dist=34988   \n",
       "189558                    Eucgr.L02107.v2.0                           NaN   \n",
       "189613                    Eucgr.L02145.v2.0                           NaN   \n",
       "189803  Eucgr.L00910.v2.0;Eucgr.L00913.v2.0         dist=38297;dist=23593   \n",
       "190139                    Eucgr.L02476.v2.0                           NaN   \n",
       "\n",
       "       ExonicFunc.refGene                                   AAChange.refGene  \n",
       "87                    NaN                                                NaN  \n",
       "95                    NaN                                                NaN  \n",
       "463     nonsynonymous SNV  Eucgr.A00810.v2.0:Eucgr.A00810.1.v2.0:exon1:c....  \n",
       "495     nonsynonymous SNV  Eucgr.A00807.v2.0:Eucgr.A00807.1.v2.0:exon1:c....  \n",
       "512     nonsynonymous SNV  Eucgr.A00807.v2.0:Eucgr.A00807.1.v2.0:exon1:c....  \n",
       "...                   ...                                                ...  \n",
       "189536                NaN                                                NaN  \n",
       "189558     synonymous SNV  Eucgr.L02107.v2.0:Eucgr.L02107.1.v2.0:exon1:c....  \n",
       "189613     synonymous SNV  Eucgr.L02145.v2.0:Eucgr.L02145.1.v2.0:exon2:c....  \n",
       "189803                NaN                                                NaN  \n",
       "190139  nonsynonymous SNV  Eucgr.L02476.v2.0:Eucgr.L02476.1.v2.0:exon5:c....  \n",
       "\n",
       "[2420 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_xylem_spec = annotated.iloc[xylem_spec.index]\n",
    "annotated_xylem_spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87              UTR5\n",
       "95              UTR3\n",
       "463           exonic\n",
       "495           exonic\n",
       "512           exonic\n",
       "             ...    \n",
       "189536    intergenic\n",
       "189558        exonic\n",
       "189613        exonic\n",
       "189803    intergenic\n",
       "190139        exonic\n",
       "Name: Func.refGene, Length: 2420, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_xylem_spec[\"Func.refGene\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Count outcome of Eucalyptus xylem-specific editing sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "func = defaultdict(int)\n",
    "exonic = defaultdict(int)\n",
    "for idx, row in annotated_xylem_spec.iterrows():\n",
    "    func[row[\"Func.refGene\"]] += 1\n",
    "    if row[\"Func.refGene\"] == \"exonic\":\n",
    "        exonic[row[\"ExonicFunc.refGene\"]] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"UTR5\": 194,\n",
      "    \"UTR3\": 242,\n",
      "    \"exonic\": 1401,\n",
      "    \"intergenic\": 353,\n",
      "    \"upstream\": 44,\n",
      "    \"intronic\": 105,\n",
      "    \"downstream\": 44,\n",
      "    \"UTR5;UTR3\": 2,\n",
      "    \"splicing\": 30,\n",
      "    \"upstream;downstream\": 3,\n",
      "    \"exonic;splicing\": 2\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(func, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"nonsynonymous SNV\": 854,\n",
      "    \"synonymous SNV\": 494,\n",
      "    \"stopgain\": 50,\n",
      "    \"stoploss\": 3\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(exonic, indent=4))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OLD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "Configs"
    ]
   },
   "outputs": [],
   "source": [
    "from matplotlib.pyplot import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "from collections import Counter\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "\n",
    "WORKDIR = Path(\"/home/b05b01002/HDD3/project_RE\")\n",
    "MYDB_PATH = Path(\"mydb\")\n",
    "FAI = {\n",
    "    \"Ptr\": Path(\"genomic_data/Ptr/Ptr.fa.fai\"),\n",
    "    \"Egr\": Path(\"genomic_data/Egr/Egr.fa.fai\")\n",
    "}\n",
    "ARCHIVE_PATH = Path(\"pyout/archive-20221207\")\n",
    "FILTERED_PATH = {\n",
    "    \"Ptr\": Path(\"shout/filtered/Ptr\"),\n",
    "    \"Egr\": Path(\"shout/filtered/Egr\")\n",
    "}\n",
    "SAMPLES = {\n",
    "    \"Ptr\": [\n",
    "        \"ptr-SAMN04535869-xr.txt\", \"ptr-SAMN04535870-xr.txt\", \"ptr-SAMN04535936-xr.txt\", \"ptr-SAMN04535937-xr.txt\", \"ptr-SAMN04535946-xr.txt\", \"ptr-SAMN04535958-xr.txt\", \"ptr-SAMN04535959-xr.txt\", \"ptr-SAMN04535960-xr.txt\", \"ptr-SAMN04536034-xr.txt\", \"ptr-SAMN04536035-xr.txt\", \"ptr-SAMN04536036-xr.txt\", \"ptr-SAMN04536096-xr.txt\", \"ptr-SAMN04536100-xr.txt\", \"ptr-SAMN04536101-xr.txt\", \"ptr-SAMN04536102-xr.txt\", \"ptr-SAMN04536244-xr.txt\", \"ptr-SAMN04536245-xr.txt\", \"ptr-SAMN04536246-xr.txt\", \"ptr-SAMN04569083-xr.txt\", \"ptr-SAMN04569084-xr.txt\", \"ptr-SAMN04569085-xr.txt\", \"ptr-SAMN04569086-xr.txt\", \"ptr-SAMN04939121-sr.txt\", \"ptr-SAMN04939122-sr.txt\", \"ptr-SAMN04939123-sr.txt\", \"ptr-SAMN04939124-lr.txt\", \"ptr-SAMN04939125-lr.txt\", \"ptr-SAMN04939126-lr.txt\", \"ptr-SAMN04939127-pr.txt\", \"ptr-SAMN04939128-pr.txt\", \"ptr-SAMN04939129-pr.txt\", \"ptr-SAMN04939134-xr.txt\", \"ptr-SAMN04939135-xr.txt\", \n",
    "        \"ptr-tsp1-xr.txt\", \"ptr-tsp2-xr.txt\", \"ptr-tsp3-xr.txt\", \"ptr-tsp4-xr.txt\", \"ptr-tsp5-xr.txt\", \"ptr-tsp6-xr.txt\", \n",
    "        \"ptr-tsp1-pr.txt\", \"ptr-tsp2-pr.txt\", \"ptr-tsp3-pr.txt\", \"ptr-tsp4-pr.txt\", \"ptr-tsp5-pr.txt\", \"ptr-tsp6-pr.txt\", \n",
    "        \"ptr-tsp1-lr.txt\", \"ptr-tsp2-lr.txt\", \"ptr-tsp3-lr.txt\", \"ptr-tsp4-lr.txt\", \"ptr-tsp5-lr.txt\", \"ptr-tsp6-lr.txt\", \n",
    "        \"ptr-tsp1-sr.txt\", \"ptr-tsp2-sr.txt\", \"ptr-tsp3-sr.txt\", \"ptr-tsp4-sr.txt\", \"ptr-tsp5-sr.txt\", \"ptr-tsp6-sr.txt\", \n",
    "    ],\n",
    "    \"Egr\": [\n",
    "        \"egr-tsp4-xr.txt\", \"egr-tsp5-xr.txt\", \"egr-tsp7-xr.txt\",\n",
    "        \"egr-tsp4-pr.txt\", \"egr-tsp5-pr.txt\", \"egr-tsp7-pr.txt\",\n",
    "        \"egr-tsp4-lr.txt\", \"egr-tsp5-lr.txt\", \"egr-tsp7-lr.txt\",\n",
    "        \"egr-tsp4-sr.txt\", \"egr-tsp5-sr.txt\", \"egr-tsp7-sr.txt\"        \n",
    "    ]\n",
    "}\n",
    "REDIT_HEADER = [\n",
    "    \"Region\", \"Position\", \"Reference\", \"Strand\", \n",
    "    \"Coverage-q25\", \"MeanQ\", \"BaseCount[A,C,G,T]\", \"AllSubs\", \"Frequency\",\n",
    "    \"gCoverage-q25\", \"gMeanQ\", \"gBaseCount[A,C,G,T]\", \"gAllSubs\", \"gFrequency\"\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [STEP0] Descriptive analysis on the RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "SPECIES = \"egr\" # ptr or egr\n",
    "PATH_COVERAGE = Path(\"shout/samtools_out/coverage/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>#rname</th>\n",
       "      <th>startpos</th>\n",
       "      <th>endpos</th>\n",
       "      <th>numreads</th>\n",
       "      <th>covbases</th>\n",
       "      <th>coverage</th>\n",
       "      <th>meandepth</th>\n",
       "      <th>meanbaseq</th>\n",
       "      <th>meanmapq</th>\n",
       "      <th>sample</th>\n",
       "      <th>source</th>\n",
       "      <th>level</th>\n",
       "      <th>meandepth_nonzero</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">egr-tsp7-sr</th>\n",
       "      <th>0</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>1</td>\n",
       "      <td>44965197</td>\n",
       "      <td>3149488</td>\n",
       "      <td>7104870</td>\n",
       "      <td>15.8008</td>\n",
       "      <td>10.35780</td>\n",
       "      <td>39.5</td>\n",
       "      <td>244.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>65.552293</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr02</td>\n",
       "      <td>1</td>\n",
       "      <td>59529170</td>\n",
       "      <td>4050082</td>\n",
       "      <td>10051556</td>\n",
       "      <td>16.8851</td>\n",
       "      <td>10.04420</td>\n",
       "      <td>39.6</td>\n",
       "      <td>235.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>59.485605</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>1</td>\n",
       "      <td>83952244</td>\n",
       "      <td>3156204</td>\n",
       "      <td>9374961</td>\n",
       "      <td>11.1670</td>\n",
       "      <td>5.52779</td>\n",
       "      <td>39.6</td>\n",
       "      <td>237.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>49.501046</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr04</td>\n",
       "      <td>1</td>\n",
       "      <td>41160059</td>\n",
       "      <td>2224894</td>\n",
       "      <td>5617211</td>\n",
       "      <td>13.6472</td>\n",
       "      <td>7.98076</td>\n",
       "      <td>39.6</td>\n",
       "      <td>244.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>58.478941</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>1</td>\n",
       "      <td>76243064</td>\n",
       "      <td>2683544</td>\n",
       "      <td>7932138</td>\n",
       "      <td>10.4038</td>\n",
       "      <td>5.16952</td>\n",
       "      <td>39.5</td>\n",
       "      <td>238.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>49.689005</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">egr-tsp5-lr</th>\n",
       "      <th>4938</th>\n",
       "      <td>scaffold_5364</td>\n",
       "      <td>1</td>\n",
       "      <td>1005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NaN</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4939</th>\n",
       "      <td>scaffold_5365</td>\n",
       "      <td>1</td>\n",
       "      <td>1934</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NaN</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4940</th>\n",
       "      <td>scaffold_5406</td>\n",
       "      <td>1</td>\n",
       "      <td>1745</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NaN</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4941</th>\n",
       "      <td>scaffold_5547</td>\n",
       "      <td>1</td>\n",
       "      <td>4138</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NaN</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4942</th>\n",
       "      <td>scaffold_5616</td>\n",
       "      <td>1</td>\n",
       "      <td>6129</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NaN</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59316 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         #rname  startpos    endpos  numreads  covbases  \\\n",
       "egr-tsp7-sr 0             Chr01         1  44965197   3149488   7104870   \n",
       "            1             Chr02         1  59529170   4050082  10051556   \n",
       "            2             Chr03         1  83952244   3156204   9374961   \n",
       "            3             Chr04         1  41160059   2224894   5617211   \n",
       "            4             Chr05         1  76243064   2683544   7932138   \n",
       "...                         ...       ...       ...       ...       ...   \n",
       "egr-tsp5-lr 4938  scaffold_5364         1      1005         0         0   \n",
       "            4939  scaffold_5365         1      1934         0         0   \n",
       "            4940  scaffold_5406         1      1745         0         0   \n",
       "            4941  scaffold_5547         1      4138         0         0   \n",
       "            4942  scaffold_5616         1      6129         0         0   \n",
       "\n",
       "                  coverage  meandepth  meanbaseq  meanmapq       sample  \\\n",
       "egr-tsp7-sr 0      15.8008   10.35780       39.5     244.0  egr-tsp7-sr   \n",
       "            1      16.8851   10.04420       39.6     235.0  egr-tsp7-sr   \n",
       "            2      11.1670    5.52779       39.6     237.0  egr-tsp7-sr   \n",
       "            3      13.6472    7.98076       39.6     244.0  egr-tsp7-sr   \n",
       "            4      10.4038    5.16952       39.5     238.0  egr-tsp7-sr   \n",
       "...                    ...        ...        ...       ...          ...   \n",
       "egr-tsp5-lr 4938    0.0000    0.00000        0.0       0.0  egr-tsp5-lr   \n",
       "            4939    0.0000    0.00000        0.0       0.0  egr-tsp5-lr   \n",
       "            4940    0.0000    0.00000        0.0       0.0  egr-tsp5-lr   \n",
       "            4941    0.0000    0.00000        0.0       0.0  egr-tsp5-lr   \n",
       "            4942    0.0000    0.00000        0.0       0.0  egr-tsp5-lr   \n",
       "\n",
       "                   source       level  meandepth_nonzero tissue  \n",
       "egr-tsp7-sr 0     lab1136  chromosome          65.552293      s  \n",
       "            1     lab1136  chromosome          59.485605      s  \n",
       "            2     lab1136  chromosome          49.501046      s  \n",
       "            3     lab1136  chromosome          58.478941      s  \n",
       "            4     lab1136  chromosome          49.689005      s  \n",
       "...                   ...         ...                ...    ...  \n",
       "egr-tsp5-lr 4938  lab1136    scaffold                NaN      l  \n",
       "            4939  lab1136    scaffold                NaN      l  \n",
       "            4940  lab1136    scaffold                NaN      l  \n",
       "            4941  lab1136    scaffold                NaN      l  \n",
       "            4942  lab1136    scaffold                NaN      l  \n",
       "\n",
       "[59316 rows x 14 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_dict = {}\n",
    "for f in glob.glob(str(PATH_COVERAGE / f\"{SPECIES}*.coverage\")):\n",
    "    cov_dict[f.split(\"/\")[-1].split(\".\")[0]] = pd.read_table(\n",
    "        f,\n",
    "        sep=\"\\t\",\n",
    "        header=0\n",
    "    )   \n",
    "result = pd.concat(\n",
    "    [v for k, v in cov_dict.items()],\n",
    "    keys=list(cov_dict.keys())\n",
    ")\n",
    "result[\"sample\"] = [i[0] for i in result.index]\n",
    "result[\"source\"] = [\"lab1136\" if \"tsp\" in i else \"SRA\" for i in result[\"sample\"]]\n",
    "result[\"level\"] = [\"chromosome\" if \"Chr\" in i else \"scaffold\" for i in result[\"#rname\"]]\n",
    "result[\"meandepth_nonzero\"] = result[\"meandepth\"] * result[\"endpos\"] / result[\"covbases\"]\n",
    "result[\"tissue\"] = [i[-2] for i in result[\"sample\"]]\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the chromosome with numreads > 5000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>#rname</th>\n",
       "      <th>startpos</th>\n",
       "      <th>endpos</th>\n",
       "      <th>numreads</th>\n",
       "      <th>covbases</th>\n",
       "      <th>coverage</th>\n",
       "      <th>meandepth</th>\n",
       "      <th>meanbaseq</th>\n",
       "      <th>meanmapq</th>\n",
       "      <th>sample</th>\n",
       "      <th>source</th>\n",
       "      <th>level</th>\n",
       "      <th>meandepth_nonzero</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">egr-tsp7-sr</th>\n",
       "      <th>0</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>1</td>\n",
       "      <td>44965197</td>\n",
       "      <td>3149488</td>\n",
       "      <td>7104870</td>\n",
       "      <td>15.8008</td>\n",
       "      <td>10.35780</td>\n",
       "      <td>39.5</td>\n",
       "      <td>244.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>65.552293</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr02</td>\n",
       "      <td>1</td>\n",
       "      <td>59529170</td>\n",
       "      <td>4050082</td>\n",
       "      <td>10051556</td>\n",
       "      <td>16.8851</td>\n",
       "      <td>10.04420</td>\n",
       "      <td>39.6</td>\n",
       "      <td>235.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>59.485605</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr03</td>\n",
       "      <td>1</td>\n",
       "      <td>83952244</td>\n",
       "      <td>3156204</td>\n",
       "      <td>9374961</td>\n",
       "      <td>11.1670</td>\n",
       "      <td>5.52779</td>\n",
       "      <td>39.6</td>\n",
       "      <td>237.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>49.501046</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr04</td>\n",
       "      <td>1</td>\n",
       "      <td>41160059</td>\n",
       "      <td>2224894</td>\n",
       "      <td>5617211</td>\n",
       "      <td>13.6472</td>\n",
       "      <td>7.98076</td>\n",
       "      <td>39.6</td>\n",
       "      <td>244.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>58.478941</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>1</td>\n",
       "      <td>76243064</td>\n",
       "      <td>2683544</td>\n",
       "      <td>7932138</td>\n",
       "      <td>10.4038</td>\n",
       "      <td>5.16952</td>\n",
       "      <td>39.5</td>\n",
       "      <td>238.0</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>49.689005</td>\n",
       "      <td>s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">egr-tsp5-lr</th>\n",
       "      <th>6</th>\n",
       "      <td>Chr07</td>\n",
       "      <td>1</td>\n",
       "      <td>54830020</td>\n",
       "      <td>2266472</td>\n",
       "      <td>6375919</td>\n",
       "      <td>11.6285</td>\n",
       "      <td>6.09150</td>\n",
       "      <td>39.6</td>\n",
       "      <td>242.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>52.384145</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Chr08</td>\n",
       "      <td>1</td>\n",
       "      <td>72515979</td>\n",
       "      <td>3344514</td>\n",
       "      <td>8922577</td>\n",
       "      <td>12.3043</td>\n",
       "      <td>6.80208</td>\n",
       "      <td>39.5</td>\n",
       "      <td>243.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>55.282178</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Chr09</td>\n",
       "      <td>1</td>\n",
       "      <td>39307835</td>\n",
       "      <td>2843674</td>\n",
       "      <td>5464825</td>\n",
       "      <td>13.9026</td>\n",
       "      <td>10.69780</td>\n",
       "      <td>39.5</td>\n",
       "      <td>248.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>76.948001</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Chr10</td>\n",
       "      <td>1</td>\n",
       "      <td>37777128</td>\n",
       "      <td>3555808</td>\n",
       "      <td>6302098</td>\n",
       "      <td>16.6823</td>\n",
       "      <td>13.95300</td>\n",
       "      <td>39.5</td>\n",
       "      <td>240.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>83.639491</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>1</td>\n",
       "      <td>44836791</td>\n",
       "      <td>3391794</td>\n",
       "      <td>6943698</td>\n",
       "      <td>15.4866</td>\n",
       "      <td>11.19230</td>\n",
       "      <td>39.5</td>\n",
       "      <td>242.0</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "      <td>lab1136</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>72.270830</td>\n",
       "      <td>l</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               #rname  startpos    endpos  numreads  covbases  coverage  \\\n",
       "egr-tsp7-sr 0   Chr01         1  44965197   3149488   7104870   15.8008   \n",
       "            1   Chr02         1  59529170   4050082  10051556   16.8851   \n",
       "            2   Chr03         1  83952244   3156204   9374961   11.1670   \n",
       "            3   Chr04         1  41160059   2224894   5617211   13.6472   \n",
       "            4   Chr05         1  76243064   2683544   7932138   10.4038   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "egr-tsp5-lr 6   Chr07         1  54830020   2266472   6375919   11.6285   \n",
       "            7   Chr08         1  72515979   3344514   8922577   12.3043   \n",
       "            8   Chr09         1  39307835   2843674   5464825   13.9026   \n",
       "            9   Chr10         1  37777128   3555808   6302098   16.6823   \n",
       "            10  Chr11         1  44836791   3391794   6943698   15.4866   \n",
       "\n",
       "                meandepth  meanbaseq  meanmapq       sample   source  \\\n",
       "egr-tsp7-sr 0    10.35780       39.5     244.0  egr-tsp7-sr  lab1136   \n",
       "            1    10.04420       39.6     235.0  egr-tsp7-sr  lab1136   \n",
       "            2     5.52779       39.6     237.0  egr-tsp7-sr  lab1136   \n",
       "            3     7.98076       39.6     244.0  egr-tsp7-sr  lab1136   \n",
       "            4     5.16952       39.5     238.0  egr-tsp7-sr  lab1136   \n",
       "...                   ...        ...       ...          ...      ...   \n",
       "egr-tsp5-lr 6     6.09150       39.6     242.0  egr-tsp5-lr  lab1136   \n",
       "            7     6.80208       39.5     243.0  egr-tsp5-lr  lab1136   \n",
       "            8    10.69780       39.5     248.0  egr-tsp5-lr  lab1136   \n",
       "            9    13.95300       39.5     240.0  egr-tsp5-lr  lab1136   \n",
       "            10   11.19230       39.5     242.0  egr-tsp5-lr  lab1136   \n",
       "\n",
       "                     level  meandepth_nonzero tissue  \n",
       "egr-tsp7-sr 0   chromosome          65.552293      s  \n",
       "            1   chromosome          59.485605      s  \n",
       "            2   chromosome          49.501046      s  \n",
       "            3   chromosome          58.478941      s  \n",
       "            4   chromosome          49.689005      s  \n",
       "...                    ...                ...    ...  \n",
       "egr-tsp5-lr 6   chromosome          52.384145      l  \n",
       "            7   chromosome          55.282178      l  \n",
       "            8   chromosome          76.948001      l  \n",
       "            9   chromosome          83.639491      l  \n",
       "            10  chromosome          72.270830      l  \n",
       "\n",
       "[132 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = result[result[\"level\"] == \"chromosome\"]\n",
    "result\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Summary by sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numreads</th>\n",
       "      <th>covbases</th>\n",
       "      <th>meandepth_nonzero</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535869-xr</th>\n",
       "      <td>5748430</td>\n",
       "      <td>44742469</td>\n",
       "      <td>6.350529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535870-xr</th>\n",
       "      <td>5460358</td>\n",
       "      <td>44306154</td>\n",
       "      <td>6.27316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535936-xr</th>\n",
       "      <td>18295547</td>\n",
       "      <td>64608436</td>\n",
       "      <td>18.365092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535937-xr</th>\n",
       "      <td>26375581</td>\n",
       "      <td>70522306</td>\n",
       "      <td>25.87479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535946-xr</th>\n",
       "      <td>21595771</td>\n",
       "      <td>66371188</td>\n",
       "      <td>24.052938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535958-xr</th>\n",
       "      <td>7688969</td>\n",
       "      <td>48079015</td>\n",
       "      <td>8.217983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535959-xr</th>\n",
       "      <td>8778284</td>\n",
       "      <td>47048253</td>\n",
       "      <td>8.776217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535960-xr</th>\n",
       "      <td>4445267</td>\n",
       "      <td>37669085</td>\n",
       "      <td>5.949551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536034-xr</th>\n",
       "      <td>9605491</td>\n",
       "      <td>59656379</td>\n",
       "      <td>15.410921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536035-xr</th>\n",
       "      <td>19253284</td>\n",
       "      <td>70270870</td>\n",
       "      <td>24.267515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536036-xr</th>\n",
       "      <td>25844454</td>\n",
       "      <td>75895031</td>\n",
       "      <td>29.489796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536096-xr</th>\n",
       "      <td>5610071</td>\n",
       "      <td>44204074</td>\n",
       "      <td>6.361258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536100-xr</th>\n",
       "      <td>82709924</td>\n",
       "      <td>83272431</td>\n",
       "      <td>59.117729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536101-xr</th>\n",
       "      <td>24075231</td>\n",
       "      <td>67808994</td>\n",
       "      <td>22.975969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536102-xr</th>\n",
       "      <td>13619481</td>\n",
       "      <td>61224675</td>\n",
       "      <td>16.118417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536244-xr</th>\n",
       "      <td>10826547</td>\n",
       "      <td>55721995</td>\n",
       "      <td>13.667138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536245-xr</th>\n",
       "      <td>10970887</td>\n",
       "      <td>59048042</td>\n",
       "      <td>13.588106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04536246-xr</th>\n",
       "      <td>4860961</td>\n",
       "      <td>42205700</td>\n",
       "      <td>8.176612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04569083-xr</th>\n",
       "      <td>9438545</td>\n",
       "      <td>57585417</td>\n",
       "      <td>12.207496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04569084-xr</th>\n",
       "      <td>16114122</td>\n",
       "      <td>61438060</td>\n",
       "      <td>19.217636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04569085-xr</th>\n",
       "      <td>13714156</td>\n",
       "      <td>59133337</td>\n",
       "      <td>16.969559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04569086-xr</th>\n",
       "      <td>14158811</td>\n",
       "      <td>62013629</td>\n",
       "      <td>16.692314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939121-sr</th>\n",
       "      <td>17005288</td>\n",
       "      <td>55717539</td>\n",
       "      <td>22.958974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939122-sr</th>\n",
       "      <td>9018799</td>\n",
       "      <td>45938027</td>\n",
       "      <td>15.513662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939123-sr</th>\n",
       "      <td>9968997</td>\n",
       "      <td>44666561</td>\n",
       "      <td>18.372028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939124-lr</th>\n",
       "      <td>12507056</td>\n",
       "      <td>43438855</td>\n",
       "      <td>19.048854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939125-lr</th>\n",
       "      <td>14652429</td>\n",
       "      <td>51532006</td>\n",
       "      <td>18.783229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939126-lr</th>\n",
       "      <td>15676812</td>\n",
       "      <td>47305662</td>\n",
       "      <td>21.508571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939127-pr</th>\n",
       "      <td>12416337</td>\n",
       "      <td>43660997</td>\n",
       "      <td>23.956149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939128-pr</th>\n",
       "      <td>12939408</td>\n",
       "      <td>47907301</td>\n",
       "      <td>21.110552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939129-pr</th>\n",
       "      <td>5986008</td>\n",
       "      <td>37810976</td>\n",
       "      <td>14.198789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939134-xr</th>\n",
       "      <td>14015473</td>\n",
       "      <td>56057040</td>\n",
       "      <td>18.516134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04939135-xr</th>\n",
       "      <td>12422115</td>\n",
       "      <td>51303661</td>\n",
       "      <td>17.880482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp1-xr</th>\n",
       "      <td>19961634</td>\n",
       "      <td>63868180</td>\n",
       "      <td>31.983116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp2-xr</th>\n",
       "      <td>21638134</td>\n",
       "      <td>64072914</td>\n",
       "      <td>35.514354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp3-xr</th>\n",
       "      <td>27016582</td>\n",
       "      <td>71080527</td>\n",
       "      <td>34.62697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp4-xr</th>\n",
       "      <td>24316720</td>\n",
       "      <td>66836332</td>\n",
       "      <td>37.808471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp5-xr</th>\n",
       "      <td>18930746</td>\n",
       "      <td>63765032</td>\n",
       "      <td>30.649492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp6-xr</th>\n",
       "      <td>23299980</td>\n",
       "      <td>67274860</td>\n",
       "      <td>35.241708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp1-pr</th>\n",
       "      <td>19217556</td>\n",
       "      <td>62851489</td>\n",
       "      <td>32.375693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp2-pr</th>\n",
       "      <td>17372100</td>\n",
       "      <td>56612869</td>\n",
       "      <td>31.875262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp3-pr</th>\n",
       "      <td>23911454</td>\n",
       "      <td>64574422</td>\n",
       "      <td>37.429086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp4-pr</th>\n",
       "      <td>25714922</td>\n",
       "      <td>63887602</td>\n",
       "      <td>40.241731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp5-pr</th>\n",
       "      <td>24985434</td>\n",
       "      <td>59951266</td>\n",
       "      <td>42.92417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp6-pr</th>\n",
       "      <td>30661076</td>\n",
       "      <td>62419468</td>\n",
       "      <td>50.507992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp1-lr</th>\n",
       "      <td>26478498</td>\n",
       "      <td>65353490</td>\n",
       "      <td>50.510786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp2-lr</th>\n",
       "      <td>25763526</td>\n",
       "      <td>66123884</td>\n",
       "      <td>46.018073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp3-lr</th>\n",
       "      <td>21304522</td>\n",
       "      <td>63053796</td>\n",
       "      <td>43.888571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp4-lr</th>\n",
       "      <td>22970774</td>\n",
       "      <td>70066443</td>\n",
       "      <td>42.928742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp5-lr</th>\n",
       "      <td>24486778</td>\n",
       "      <td>68844608</td>\n",
       "      <td>43.04514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp6-lr</th>\n",
       "      <td>22427870</td>\n",
       "      <td>65134969</td>\n",
       "      <td>44.11481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp1-sr</th>\n",
       "      <td>19223338</td>\n",
       "      <td>61432635</td>\n",
       "      <td>37.95041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp2-sr</th>\n",
       "      <td>19176302</td>\n",
       "      <td>61810139</td>\n",
       "      <td>36.628659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp3-sr</th>\n",
       "      <td>22192680</td>\n",
       "      <td>60796864</td>\n",
       "      <td>42.123069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp4-sr</th>\n",
       "      <td>29649570</td>\n",
       "      <td>65018223</td>\n",
       "      <td>54.418272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp5-sr</th>\n",
       "      <td>25516776</td>\n",
       "      <td>65880432</td>\n",
       "      <td>44.121489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-tsp6-sr</th>\n",
       "      <td>35732706</td>\n",
       "      <td>70630343</td>\n",
       "      <td>58.113576</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     numreads  covbases meandepth_nonzero\n",
       "ptr-SAMN04535869-xr   5748430  44742469          6.350529\n",
       "ptr-SAMN04535870-xr   5460358  44306154           6.27316\n",
       "ptr-SAMN04535936-xr  18295547  64608436         18.365092\n",
       "ptr-SAMN04535937-xr  26375581  70522306          25.87479\n",
       "ptr-SAMN04535946-xr  21595771  66371188         24.052938\n",
       "ptr-SAMN04535958-xr   7688969  48079015          8.217983\n",
       "ptr-SAMN04535959-xr   8778284  47048253          8.776217\n",
       "ptr-SAMN04535960-xr   4445267  37669085          5.949551\n",
       "ptr-SAMN04536034-xr   9605491  59656379         15.410921\n",
       "ptr-SAMN04536035-xr  19253284  70270870         24.267515\n",
       "ptr-SAMN04536036-xr  25844454  75895031         29.489796\n",
       "ptr-SAMN04536096-xr   5610071  44204074          6.361258\n",
       "ptr-SAMN04536100-xr  82709924  83272431         59.117729\n",
       "ptr-SAMN04536101-xr  24075231  67808994         22.975969\n",
       "ptr-SAMN04536102-xr  13619481  61224675         16.118417\n",
       "ptr-SAMN04536244-xr  10826547  55721995         13.667138\n",
       "ptr-SAMN04536245-xr  10970887  59048042         13.588106\n",
       "ptr-SAMN04536246-xr   4860961  42205700          8.176612\n",
       "ptr-SAMN04569083-xr   9438545  57585417         12.207496\n",
       "ptr-SAMN04569084-xr  16114122  61438060         19.217636\n",
       "ptr-SAMN04569085-xr  13714156  59133337         16.969559\n",
       "ptr-SAMN04569086-xr  14158811  62013629         16.692314\n",
       "ptr-SAMN04939121-sr  17005288  55717539         22.958974\n",
       "ptr-SAMN04939122-sr   9018799  45938027         15.513662\n",
       "ptr-SAMN04939123-sr   9968997  44666561         18.372028\n",
       "ptr-SAMN04939124-lr  12507056  43438855         19.048854\n",
       "ptr-SAMN04939125-lr  14652429  51532006         18.783229\n",
       "ptr-SAMN04939126-lr  15676812  47305662         21.508571\n",
       "ptr-SAMN04939127-pr  12416337  43660997         23.956149\n",
       "ptr-SAMN04939128-pr  12939408  47907301         21.110552\n",
       "ptr-SAMN04939129-pr   5986008  37810976         14.198789\n",
       "ptr-SAMN04939134-xr  14015473  56057040         18.516134\n",
       "ptr-SAMN04939135-xr  12422115  51303661         17.880482\n",
       "ptr-tsp1-xr          19961634  63868180         31.983116\n",
       "ptr-tsp2-xr          21638134  64072914         35.514354\n",
       "ptr-tsp3-xr          27016582  71080527          34.62697\n",
       "ptr-tsp4-xr          24316720  66836332         37.808471\n",
       "ptr-tsp5-xr          18930746  63765032         30.649492\n",
       "ptr-tsp6-xr          23299980  67274860         35.241708\n",
       "ptr-tsp1-pr          19217556  62851489         32.375693\n",
       "ptr-tsp2-pr          17372100  56612869         31.875262\n",
       "ptr-tsp3-pr          23911454  64574422         37.429086\n",
       "ptr-tsp4-pr          25714922  63887602         40.241731\n",
       "ptr-tsp5-pr          24985434  59951266          42.92417\n",
       "ptr-tsp6-pr          30661076  62419468         50.507992\n",
       "ptr-tsp1-lr          26478498  65353490         50.510786\n",
       "ptr-tsp2-lr          25763526  66123884         46.018073\n",
       "ptr-tsp3-lr          21304522  63053796         43.888571\n",
       "ptr-tsp4-lr          22970774  70066443         42.928742\n",
       "ptr-tsp5-lr          24486778  68844608          43.04514\n",
       "ptr-tsp6-lr          22427870  65134969          44.11481\n",
       "ptr-tsp1-sr          19223338  61432635          37.95041\n",
       "ptr-tsp2-sr          19176302  61810139         36.628659\n",
       "ptr-tsp3-sr          22192680  60796864         42.123069\n",
       "ptr-tsp4-sr          29649570  65018223         54.418272\n",
       "ptr-tsp5-sr          25516776  65880432         44.121489\n",
       "ptr-tsp6-sr          35732706  70630343         58.113576"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_ptr = pd.DataFrame(columns=[\"numreads\", \"covbases\", \"meandepth_nonzero\"])\n",
    "for sample in SAMPLES[\"Ptr\"]:\n",
    "    sample = sample.strip(\".txt\")\n",
    "    summary_ptr.loc[sample, \"numreads\"] = result.loc[sample, \"numreads\"].sum()\n",
    "    summary_ptr.loc[sample, \"covbases\"] = result.loc[sample, \"covbases\"].sum()\n",
    "    summary_ptr.loc[sample, \"meandepth_nonzero\"] = result.loc[sample, \"meandepth_nonzero\"].mean()\n",
    "summary_ptr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numreads</th>\n",
       "      <th>covbases</th>\n",
       "      <th>meandepth_nonzero</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>egr-tsp4-xr</th>\n",
       "      <td>35226858</td>\n",
       "      <td>67559164</td>\n",
       "      <td>25.267239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-xr</th>\n",
       "      <td>35946838</td>\n",
       "      <td>65060426</td>\n",
       "      <td>24.59622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-xr</th>\n",
       "      <td>38726238</td>\n",
       "      <td>64514799</td>\n",
       "      <td>25.120462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp4-pr</th>\n",
       "      <td>30347988</td>\n",
       "      <td>60856536</td>\n",
       "      <td>27.124655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-pr</th>\n",
       "      <td>31894920</td>\n",
       "      <td>66343622</td>\n",
       "      <td>25.662697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-pr</th>\n",
       "      <td>36030480</td>\n",
       "      <td>65924619</td>\n",
       "      <td>28.751322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp4-lr</th>\n",
       "      <td>31266106</td>\n",
       "      <td>80675688</td>\n",
       "      <td>20.254982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-lr</th>\n",
       "      <td>34412962</td>\n",
       "      <td>82407076</td>\n",
       "      <td>24.492633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-lr</th>\n",
       "      <td>33493680</td>\n",
       "      <td>83408015</td>\n",
       "      <td>18.4625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp4-sr</th>\n",
       "      <td>30666196</td>\n",
       "      <td>89472234</td>\n",
       "      <td>16.808683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-sr</th>\n",
       "      <td>34497068</td>\n",
       "      <td>91131047</td>\n",
       "      <td>19.47585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-sr</th>\n",
       "      <td>37277448</td>\n",
       "      <td>93049733</td>\n",
       "      <td>17.647047</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             numreads  covbases meandepth_nonzero\n",
       "egr-tsp4-xr  35226858  67559164         25.267239\n",
       "egr-tsp5-xr  35946838  65060426          24.59622\n",
       "egr-tsp7-xr  38726238  64514799         25.120462\n",
       "egr-tsp4-pr  30347988  60856536         27.124655\n",
       "egr-tsp5-pr  31894920  66343622         25.662697\n",
       "egr-tsp7-pr  36030480  65924619         28.751322\n",
       "egr-tsp4-lr  31266106  80675688         20.254982\n",
       "egr-tsp5-lr  34412962  82407076         24.492633\n",
       "egr-tsp7-lr  33493680  83408015           18.4625\n",
       "egr-tsp4-sr  30666196  89472234         16.808683\n",
       "egr-tsp5-sr  34497068  91131047          19.47585\n",
       "egr-tsp7-sr  37277448  93049733         17.647047"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_egr = pd.DataFrame(columns=[\"numreads\", \"covbases\", \"meandepth_nonzero\"])\n",
    "for sample in SAMPLES[\"Egr\"]:\n",
    "    sample = sample.strip(\".txt\")\n",
    "    summary_egr.loc[sample, \"numreads\"] = result.loc[sample, \"numreads\"].sum()\n",
    "    summary_egr.loc[sample, \"covbases\"] = result.loc[sample, \"covbases\"].sum()\n",
    "    summary_egr.loc[sample, \"meandepth_nonzero\"] = result.loc[sample, \"meandepth_nonzero\"].mean()\n",
    "summary_egr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numreads</th>\n",
       "      <th>covbases</th>\n",
       "      <th>meandepth_nonzero</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535869-xr</th>\n",
       "      <td>5748430</td>\n",
       "      <td>44742469</td>\n",
       "      <td>6.350529</td>\n",
       "      <td>ptr-SAMN04535869-xr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535870-xr</th>\n",
       "      <td>5460358</td>\n",
       "      <td>44306154</td>\n",
       "      <td>6.27316</td>\n",
       "      <td>ptr-SAMN04535870-xr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535936-xr</th>\n",
       "      <td>18295547</td>\n",
       "      <td>64608436</td>\n",
       "      <td>18.365092</td>\n",
       "      <td>ptr-SAMN04535936-xr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535937-xr</th>\n",
       "      <td>26375581</td>\n",
       "      <td>70522306</td>\n",
       "      <td>25.87479</td>\n",
       "      <td>ptr-SAMN04535937-xr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ptr-SAMN04535946-xr</th>\n",
       "      <td>21595771</td>\n",
       "      <td>66371188</td>\n",
       "      <td>24.052938</td>\n",
       "      <td>ptr-SAMN04535946-xr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-lr</th>\n",
       "      <td>34412962</td>\n",
       "      <td>82407076</td>\n",
       "      <td>24.492633</td>\n",
       "      <td>egr-tsp5-lr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-lr</th>\n",
       "      <td>33493680</td>\n",
       "      <td>83408015</td>\n",
       "      <td>18.4625</td>\n",
       "      <td>egr-tsp7-lr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp4-sr</th>\n",
       "      <td>30666196</td>\n",
       "      <td>89472234</td>\n",
       "      <td>16.808683</td>\n",
       "      <td>egr-tsp4-sr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp5-sr</th>\n",
       "      <td>34497068</td>\n",
       "      <td>91131047</td>\n",
       "      <td>19.47585</td>\n",
       "      <td>egr-tsp5-sr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>egr-tsp7-sr</th>\n",
       "      <td>37277448</td>\n",
       "      <td>93049733</td>\n",
       "      <td>17.647047</td>\n",
       "      <td>egr-tsp7-sr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     numreads  covbases meandepth_nonzero               sample\n",
       "ptr-SAMN04535869-xr   5748430  44742469          6.350529  ptr-SAMN04535869-xr\n",
       "ptr-SAMN04535870-xr   5460358  44306154           6.27316  ptr-SAMN04535870-xr\n",
       "ptr-SAMN04535936-xr  18295547  64608436         18.365092  ptr-SAMN04535936-xr\n",
       "ptr-SAMN04535937-xr  26375581  70522306          25.87479  ptr-SAMN04535937-xr\n",
       "ptr-SAMN04535946-xr  21595771  66371188         24.052938  ptr-SAMN04535946-xr\n",
       "...                       ...       ...               ...                  ...\n",
       "egr-tsp5-lr          34412962  82407076         24.492633          egr-tsp5-lr\n",
       "egr-tsp7-lr          33493680  83408015           18.4625          egr-tsp7-lr\n",
       "egr-tsp4-sr          30666196  89472234         16.808683          egr-tsp4-sr\n",
       "egr-tsp5-sr          34497068  91131047          19.47585          egr-tsp5-sr\n",
       "egr-tsp7-sr          37277448  93049733         17.647047          egr-tsp7-sr\n",
       "\n",
       "[69 rows x 4 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_all = pd.concat([summary_ptr, summary_egr])\n",
    "summary_all[\"sample\"] = summary_all.index\n",
    "summary_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot: xlabel='sample', ylabel='numreads'>"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAm8AAAHACAYAAAD0lI8uAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA7FUlEQVR4nO3deXxTVf7/8XeatuleoKyFQlllK4tUFBDBUdCyDLKpKFJZvg4jCoOgiKAConVFRh1AXBB1cAPBHcEFcQaYH7Irg4ogVEQRdChrwfb8/oBcm2ahQJv0htfz8bgPyDm5935yc5O8e+69icMYYwQAAABbiAh1AQAAACg5whsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN7KwPLly9WzZ0+lpqbK4XBo0aJFpzX/pEmT5HA4vKb4+PiyKRgAANgG4a0MHDp0SC1bttRTTz11RvOPHTtWu3fv9piaNm2q/v37l3KlAADAbghvZSArK0tTp05Vnz59fPYfO3ZMd9xxh2rWrKn4+HhdeOGFWrZsmdWfkJCg6tWrW9PPP/+szZs3a+jQoUF6BAAAoLyKDHUB56LBgwfr+++/16uvvqrU1FQtXLhQV155pTZt2qSGDRt63f/ZZ59Vo0aN1LFjxxBUCwAAyhNG3oLsu+++0yuvvKI33nhDHTt2VP369TV27FhdfPHFmjNnjtf98/Pz9c9//pNRNwAAIImRt6Bbu3atjDFq1KiRR3t+fr5SUlK87v/mm2/qwIEDGjRoULBKBAAA5RjhLcgKCwvldDq1Zs0aOZ1Oj76EhASv+z/77LPq0aOHqlevHqwSAQBAOUZ4C7LWrVuroKBAe/bsOeU5bNu3b9enn36qt99+O0jVAQCA8o7wVgYOHjyorVu3Wre3b9+u9evXq1KlSmrUqJGuv/56DRo0SI899phat26tvXv36pNPPlFGRoa6detmzff888+rRo0aysrKCsXDAAAA5ZDDGGNCXUS4WbZsmS699FKv9uzsbL3wwgs6fvy4pk6dqhdffFG7du1SSkqK2rVrp8mTJysjI0PSicOrderU0aBBg3T//fcH+yEAAIByivAGAABgI3xVCAAAgI0Q3gAAAGyECxZKUWFhoX788UclJibK4XCEuhwAAFACxhgdOHBAqampiogo/+NahLdS9OOPPyotLS3UZQAAgDOQm5urWrVqhbqMUyK8laLExERJJ578pKSkEFcDAABKIi8vT2lpadbneHlHeCtF7kOlSUlJhDcAAGzGLqc8lf8DuwAAALAQ3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANhIZ6gKAkvrguW4+27OGvh/kSgAACB1G3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4O+n333/XxIkTVbduXcXGxqpevXqaMmWKCgsLQ10aAACAhe95O+mhhx7SrFmzNHfuXDVr1kxffPGFBg8erOTkZI0aNSrU5QEAAEgivFlWrlypXr16qXv37pKk9PR0vfLKK/riiy9CXBkAAMAfOGx60sUXX6yPP/5Y33zzjSRpw4YN+te//qVu3Xx/q78k5efnKy8vz2MCAAAoS4y8nTRu3Djt379fjRs3ltPpVEFBge6//34NGDDA7zw5OTmaPHlyEKsEAADnOkbeTnrttdf08ssva968eVq7dq3mzp2rRx99VHPnzvU7z/jx47V//35rys3NDWLFAADgXMTI20m333677rzzTl177bWSpIyMDO3YsUM5OTnKzs72OY/L5ZLL5QpmmQAA4BzHyNtJhw8fVkSE5+ZwOp18VQgAAChXGHk7qWfPnrr//vtVu3ZtNWvWTOvWrdO0adM0ZMiQUJcGAABgIbyd9OSTT+ruu+/WzTffrD179ig1NVV/+ctfdM8994S6NAAAAAvh7aTExERNnz5d06dPD3UpAAAAfnHOGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4Oyk9PV0Oh8NrGjFiRKhLAwAAsESGuoDyYvXq1SooKLBuf/nll+rSpYv69+8fwqoAAAA8Ed5OqlKlisftBx98UPXr11enTp1CVBEAAIA3Dpv6cOzYMb388ssaMmSIHA5HqMsBAACwMPLmw6JFi/S///1PN954Y8D75efnKz8/37qdl5dXxpUBAIBzHSNvPjz33HPKyspSampqwPvl5OQoOTnZmtLS0oJUIQAAOFcR3orZsWOHPvroIw0bNuyU9x0/frz2799vTbm5uUGoEAAAnMs4bFrMnDlzVLVqVXXv3v2U93W5XHK5XEGoCgAA4ARG3oooLCzUnDlzlJ2drchIci0AACh/CG9FfPTRR9q5c6eGDBkS6lIAAAB8YnipiK5du8oYE+oyAAAA/GLkDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhrYhdu3Zp4MCBSklJUVxcnFq1aqU1a9aEuiwAAABLZKgLKC9+++03dejQQZdeeqk++OADVa1aVd99950qVKgQ6tIAAAAshLeTHnroIaWlpWnOnDlWW3p6eugKAgAA8IHDpie9/fbbyszMVP/+/VW1alW1bt1azzzzTMB58vPzlZeX5zEBAACUJcLbSdu2bdPMmTPVsGFDffjhhxo+fLhGjhypF1980e88OTk5Sk5Otqa0tLQgVgwAAM5FDmOMCXUR5UF0dLQyMzO1YsUKq23kyJFavXq1Vq5c6XOe/Px85efnW7fz8vKUlpam/fv3KykpqcxrPtd88Fw3n+1ZQ98PciUAgHCSl5en5ORk23x+M/J2Uo0aNdS0aVOPtiZNmmjnzp1+53G5XEpKSvKYAAAAyhLh7aQOHTro66+/9mj75ptvVKdOnRBVBAAA4I3wdtLo0aO1atUqPfDAA9q6davmzZun2bNna8SIEaEuDQAAwEJ4O+mCCy7QwoUL9corr6h58+a67777NH36dF1//fWhLg0AAMDC97wV0aNHD/Xo0SPUZQAAAPjFyBsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeDtp0qRJcjgcHlP16tVDXRYAAICHyFAXUJ40a9ZMH330kXXb6XSGsBoAAABvYTHyNnfuXL333nvW7TvuuEMVKlRQ+/bttWPHjhIvJzIyUtWrV7emKlWqlEW5AAAAZywswtsDDzyg2NhYSdLKlSv11FNP6eGHH1blypU1evToEi/n22+/VWpqqurWratrr71W27ZtK6uSAQAAzkhYHDbNzc1VgwYNJEmLFi1Sv379dNNNN6lDhw7q3LlziZZx4YUX6sUXX1SjRo30888/a+rUqWrfvr2++uorpaSk+JwnPz9f+fn51u28vLyzfiwAAACBhMXIW0JCgvbt2ydJWrJkiS6//HJJUkxMjI4cOVKiZWRlZalv377KyMjQ5Zdfbh2GnTt3rt95cnJylJycbE1paWln+UgAAAACC4vw1qVLFw0bNkzDhg3TN998o+7du0uSvvrqK6Wnp5/RMuPj45WRkaFvv/3W733Gjx+v/fv3W1Nubu4ZrQsAAKCkwiK8/eMf/1C7du30yy+/aMGCBdZhzjVr1mjAgAFntMz8/Hz997//VY0aNfzex+VyKSkpyWMCAAAoS2FxzluFChX01FNPebVPnjy5xMsYO3asevbsqdq1a2vPnj2aOnWq8vLylJ2dXZqlAgAAnBXbhreNGzeW+L4tWrQ45X1++OEHDRgwQHv37lWVKlV00UUXadWqVapTp87ZlAkAAFCqbBveWrVqJYfDIWOMHA5HwPsWFBSccnmvvvpqaZUGAABQZmx7ztv27du1bds2bd++XQsWLFDdunU1Y8YMrVu3TuvWrdOMGTNUv359LViwINSlAgAAlBrbjrwVPZzZv39/PfHEE+rWrZvV1qJFC6Wlpenuu+/WVVddFYIKAQAASp9tR96K2rRpk+rWrevVXrduXW3evDkEFQEAAJSNsAhvTZo00dSpU3X06FGrLT8/X1OnTlWTJk1CWBkAAEDpsu1h06JmzZqlnj17Ki0tTS1btpQkbdiwQQ6HQ++++26IqwMAACg9YRHe2rZtq+3bt+vll1/Wli1bZIzRNddco+uuu07x8fGhLg8AAKDUhEV4k6S4uDjddNNNoS4DAACgTIVNeJOkzZs3a+fOnTp27JhH+5///OcQVQQAAFC6wiK8bdu2Tb1799amTZusL+6VZH15b0m+pBcAAMAOwuJq01GjRqlu3br6+eefFRcXp6+++krLly9XZmamli1bFuryAAAASk1YjLytXLlSn3zyiapUqaKIiAhFRETo4osvVk5OjkaOHKl169aFukQAAIBSERYjbwUFBUpISJAkVa5cWT/++KOkE7/C8PXXX4eyNAAAgFIVFiNvzZs318aNG1WvXj1deOGFevjhhxUdHa3Zs2erXr16oS4PAACg1IRFeJs4caIOHTokSZo6dap69Oihjh07KiUlRa+99lqIqwMAACg9YRHerrjiCuv/9erV0+bNm/Xrr7+qYsWK1hWnAAAA4SAsznlz27p1qz788EMdOXJElSpVCnU5AAAApS4swtu+fft02WWXqVGjRurWrZt2794tSRo2bJjGjBkT4uoAAABKT1iEt9GjRysqKko7d+5UXFyc1X7NNddo8eLFIawMAACgdIXFOW9LlizRhx9+qFq1anm0N2zYUDt27AhRVQAAAKUvLEbeDh065DHi5rZ37165XK4QVAQAAFA2wiK8XXLJJXrxxRet2w6HQ4WFhXrkkUd06aWXhrAyAACA0hUWh00feeQRde7cWV988YWOHTumO+64Q1999ZV+/fVX/fvf/w51eQAAAKUmLEbemjZtqo0bN6pt27bq0qWLDh06pD59+mjdunWqX79+qMsDAAAoNbYfeTt+/Li6du2qp59+WpMnTw51OQAAAGXK9iNvUVFR+vLLL/klBQAAcE6wfXiTpEGDBum5554LdRkAAABlzvaHTSXp2LFjevbZZ7V06VJlZmYqPj7eo3/atGkhqgwAAKB0hUV4+/LLL3X++edLkr755huPPg6nAgCAcBIW4e3TTz8NdQkAAABBERbnvAEAAJwrwmLk7ejRo3ryySf16aefas+ePSosLPToX7t2bYgqAwAAKF1hEd6GDBmipUuXql+/fmrbti3nuQEAgLAVFuHtvffe0/vvv68OHTqU2jJzcnJ01113adSoUZo+fXqpLRcAAOBshMU5bzVr1lRiYmKpLW/16tWaPXu2WrRoUWrLBAAAKA1hEd4ee+wxjRs3Tjt27DjrZR08eFDXX3+9nnnmGVWsWLEUqgMAACg9YRHeMjMzdfToUdWrV0+JiYmqVKmSx3Q6RowYoe7du+vyyy8/5X3z8/OVl5fnMQEAAJSlsDjnbcCAAdq1a5ceeOABVatW7YwvWHj11Ve1du1arV69ukT3z8nJ0eTJk89oXQAAAGciLMLbihUrtHLlSrVs2fKMl5Gbm6tRo0ZpyZIliomJKdE848eP12233WbdzsvLU1pa2hnXAAAAcCphEd4aN26sI0eOnNUy1qxZoz179qhNmzZWW0FBgZYvX66nnnpK+fn5cjqdHvO4XC65XK6zWi8AAMDpCIvw9uCDD2rMmDG6//77lZGRoaioKI/+pKSkUy7jsssu06ZNmzzaBg8erMaNG2vcuHFewQ0AACAUwiK8XXnllZJOBLCijDFyOBwqKCg45TISExPVvHlzj7b4+HilpKR4tQMAAIRKWIQ3fpgeAACcK8IivHXq1KlMlrts2bIyWS4AAKGw+5Fcr7Yat3Ohnd2ERXhbvnx5wP5LLrkkSJUAAACUrbAIb507d/ZqK/pdbyU55w0AAMAOwuIXFn777TePac+ePVq8eLEuuOACLVmyJNTlAQAAlJqwGHlLTk72auvSpYtcLpdGjx6tNWvWhKAqAACA0hcWI2/+VKlSRV9//XWoywAAACg1YTHytnHjRo/bxhjt3r1bDz744Fn9ZBYAAAgfe2bM99keM7BrkCs5O2ER3lq1aiWHwyFjjEf7RRddpOeffz5EVQEAAJS+sAhv27dv97gdERGhKlWqlPgH5gEAAOwiLMJbnTp19PHHH+vjjz/Wnj17VFhY6NHP6BsAAAgXYRHeJk+erClTpigzM1M1atTw+I43AACAcBIW4W3WrFl64YUXdMMNN4S6FAAAzik/T1/rs73a384PciXnjrAIb8eOHVP79u1DXQYAACH1w2M/+WyvNaZ6kCtBWQqL8DZs2DDNmzdPd999d6hLAQDgrH0162ef7c2GVwtyJSiPwiK8HT16VLNnz9ZHH32kFi1aKCoqyqN/2rRpIaoMAACgdIVFeNu4caNatWolSfryyy89+rh4AQAQTrY+6XtUrsGtjMqdK8IivH366aehLgEAACAowvq3TQEAAMJNWIy8AQAA+/n5ieVebdVGXnJWy9zzj7d9tlcd8eezWm55wsgbAACAjRDeAAAAbITwBgAAYCOc8wYAAMqdPU9+7LO96q2XBbmS8oeRNwAAABshvAEAANgI4Q0AAMBGCG8AAAA2QngDAACwEcIbAACAjfBVIQAAIKCfpn3ps736bc2DXAkkRt4AAABshfAGAABgIxw2BQAAZeLnv6/y2V5t1EVBriS8MPJ20syZM9WiRQslJSUpKSlJ7dq10wcffBDqsgAAADww8nZSrVq19OCDD6pBgwaSpLlz56pXr15at26dmjVrFuLqAKB86r/gK5/tb/TlfRNlZ89TvgdXqt6SFeRKQoPwdlLPnj09bt9///2aOXOmVq1aRXgDAADlBuHNh4KCAr3xxhs6dOiQ2rVr5/d++fn5ys/Pt27n5eUFozwAAErdT49967O9+piGQa4Ep8I5b0Vs2rRJCQkJcrlcGj58uBYuXKimTZv6vX9OTo6Sk5OtKS0tLYjVAgCAcxHhrYjzzjtP69ev16pVq/TXv/5V2dnZ2rx5s9/7jx8/Xvv377em3NzcIFYLAADORRw2LSI6Otq6YCEzM1OrV6/W3//+dz399NM+7+9yueRyuYJZIgDARlbO/cWrrV12lRBUgnDCyFsAxhiPc9oAAABCjZG3k+666y5lZWUpLS1NBw4c0Kuvvqply5Zp8eLFoS4NAADAQng76eeff9YNN9yg3bt3Kzk5WS1atNDixYvVpUuXUJcGAABgIbyd9Nxzz4W6BAAAgFPinDcAAAAbIbwBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI3wVSEAcIZ6zH/FZ/u7/QYEuRIA5xJG3gAAAGyEkTeb+Hnmwz7bq/31jiBXAgAlc82b23y2v9anXpArAcILI28AAAA2QngDAACwEQ6bIqiWP9PdZ/sl//dekCsB4NZnwUqf7W/2bVdm6xy5MNdn+xO908psnUC4ILwBABACa5/b47P9/KFVg1wJ7IbDpgAAADbCyBsAIKB+C9Z4tc3v2yYElQCQGHkDAACwFcIbAACAjXDYFADKQI/5b/hsf7df/yBXcm55/k3fFwEM6cNFAAgfhDeEhbeez/LZ3mvIB0GuBACAskV4A4BzQO8Fn/lsX9i3U5ArAXC2OOcNAADARghvAAAANsJhUwAoZ/48/x2vtrf79QxBJQDKI8IbAMBW/r7wJ6+2Ub2rh6ASIDQ4bAoAAGAjjLwBAM4Jry3Y67P9mr6Vg1wJcHYIbwAAnIVl//zFq63z9VVCUAnOFRw2BQAAsBHCGwAAgI1w2DQM7J4x0Wd7jZunBrkSIPz0mP+Sz/Z3+90Q5EoA4ATCGwCgXJmy8Eef7ff0Tg1yJUD5xGHTk3JycnTBBRcoMTFRVatW1VVXXaWvv/461GUBAAB4ILyd9Nlnn2nEiBFatWqVli5dqt9//11du3bVoUOHQl0aAACAhcOmJy1evNjj9pw5c1S1alWtWbNGl1xySYiqAgAA8MTImx/79++XJFWqVCnElQAAAPyBkTcfjDG67bbbdPHFF6t58+Z+75efn6/8/Hzrdl5eXjDKAwAA5zDCmw+33HKLNm7cqH/9618B75eTk6PJkycHqSoAZaXHghd8tr/b98ag1gEAJcFh02JuvfVWvf322/r0009Vq1atgPcdP3689u/fb025ublBqhIAAJyrGHk7yRijW2+9VQsXLtSyZctUt27dU87jcrnkcrmCUB0AAMAJhLeTRowYoXnz5umtt95SYmKifvrpJ0lScnKyYmNjQ1ydvfzn6R4+2y/8y7tBrgQIP73mL/bZ/la/K4NcSXh56429Ptt79a+sD1/13XfFtZXLsiTAL8LbSTNnzpQkde7c2aN9zpw5uvHGG4NfEICw1nP+Qq+2d/r1DkElAOyG8HaSMSbUJQAAAJwSFywAAADYCCNv5cgvs/7hs73K8BFBrgQAAJRXjLwBAADYCOENAADARghvAAAANsI5bwAQJq6a/5HP9kX9Lg9yJQDKEuENCBO3LfD+ktZpfX1/oSsAwL44bAoAAGAjhDcAAAAbIbwBAADYCOe8AUE2bd4VPttvu+7DIFcCALAjRt4AAABshJE3AGWi28J7vdre7z05BJUAQHhh5A0AAMBGCG8AAAA2QngDAACwEc55A3BGui263Wf7+1c9EuRKAODcwsgbAACAjTDyBtjI3a97/37pfVfz+6UAcC4hvAEIC93fnOmz/b0+fw1yJQBQtghvQDnywKu+f33hrmv59QUAwAmEN4S9N+Z4H2qUpP6DOdwIALAfLlgAAACwEcIbAACAjXDYFEC50v3Nx3y2v9dnTJArAYDyifAGwK+st27y2f5Br9lBrgQA4EZ4AxB03Rbm+Gx/v/f4Mltn9wXPerW913dYma0PAMoK57wBAADYCOENAADARghvAAAANkJ4AwAAsBHCGwAAgI1wtWkRy5cv1yOPPKI1a9Zo9+7dWrhwoa666qpQl4UQmjO3q1fb4OwlIagEAIATCG9FHDp0SC1bttTgwYPVt2/fUJeDIPjnC75/CP76G/kheABA+UR4KyIrK0tZWVmhLgMAAMAvwttZyM/PV35+vnU7Ly8vhNUAAIBzAeHtLOTk5Gjy5MmhLgM4K1lv/9ln+wd/fjvIlQAASoKrTc/C+PHjtX//fmvKzc0NdUkAACDMMfJ2Flwul1wuV6jLCIn1s3r6bG81/J0gVwIAwLmF8AaUgade9n0V6y0DuYoVAHB2CG9FHDx4UFu3brVub9++XevXr1elSpVUu3btEFYGnJ3BC6/02T6n9+IgVwIAOFuEtyK++OILXXrppdbt2267TZKUnZ2tF154IURVla3vn7jKqy195KKg1wEAAEqG8FZE586dZYwJdRkAAAB+cbUpAACAjRDeAAAAbITwBgAAYCOc83YO2PXUX73aat4yMwSVAACAs8XIGwAAgI0Q3gAAAGyE8AYAAGAjhDcAAAAbIbwBAADYCOENAADARviqEAC20f3NJ3y2v9dnZJArAYDQIbyVgb3Pvqb82FiPtip/HRiiagAAQDghvAXZL7Oe89leZfjQIFdSPn30bHevtsuHvReCSgAAKJ845w0AAMBGCG8AAAA2wmFT4AzNfukKn+033fBhkCsBAJxLGHkDAACwEUbe4Nd//9HLZ3uTEW8FuRIAAODGyBsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhrZgZM2aobt26iomJUZs2bfT555+HuiQAAAAL4a2I1157TX/72980YcIErVu3Th07dlRWVpZ27twZ6tIAAAAkEd48TJs2TUOHDtWwYcPUpEkTTZ8+XWlpaZo5c2aoSwMAAJBEeLMcO3ZMa9asUdeuXT3au3btqhUrVoSoKgAAAE+RoS6gvNi7d68KCgpUrVo1j/Zq1arpp59+8jlPfn6+8vPzrdv79++XJB04esTrvq68vBN9R7z73P2B+4767IvNy9OBI/k+++KtdR7z6stz9x097rfv4BHvPnd/oL5DZ9gnyWe/u+9wgHkPH/m91Psk6YiP/kB97v5AfUcPn1mfJOX76C9J37EAy/39sP/t+vth733H3Xf8sO/9zr1OX/1/9Pnen08s90z7fL9+/lind3+gvpIs9/jhw2fUd2Kd3v0l6zsUYJ2l33dinQdPq++P5R7w23csQN/RAH2SfPbn5cVJko74nTdGh/32RQfsOxSgT5LP/rw814m+I/77Dvroc9fqvy9WBwL1HfXXd2L7+Oq3PiuO+n4u4/LyzqgvNi9PB4763rdirXV691t9R3zPG5OXpwNHfL++zrTv2Ml1GmN89pc7BsYYY3bt2mUkmRUrVni0T5061Zx33nk+57n33nuNJCYmJiYmJqYwmHJzc4MROc4aI28nVa5cWU6n02uUbc+ePV6jcW7jx4/XbbfdZt0uLCzUr7/+qpSUFB04cEBpaWnKzc1VUlKS17x5eXl++4PdRz3hUyv1hE+t1BM+tVJP+a/VGKMDBw4oNTXVa/3lEeHtpOjoaLVp00ZLly5V7969rfalS5eqV69ePudxuVxyuVwebRUqVJAkORwOSVJSUpLPndEtUH+w+6gnfGqlnvCplXrCp1bqKd+1Jicn+113eUN4K+K2227TDTfcoMzMTLVr106zZ8/Wzp07NXz48FCXBgAAIInw5uGaa67Rvn37NGXKFO3evVvNmzfX+++/rzp16oS6NAAAAEmENy8333yzbr755rNejsvl0r333ut1WLUk/cHuo57wqZV6wqdW6gmfWqnHXrXagcMYu1wXCwAAAL6kFwAAwEYIbwAAADZCeAOA05Cenq7p06efdl9J+nF2XnjhBevrmkLdV5J+lJ2yek7LjdL8xt86deqYxx9//LT7jDGmdu3a5k9/+pOpW7eucblcpnLlyqZTp07mnXfe8eqrVKmScTgcpmbNmqZOnTrmoYceMnfccYfVr5PflFyxYkW/fQ6Hw+or2ub+f7Vq1UzlypWt28nJyV73i4iI8Pp25oiICON0Ok10dLRXn8vlMhUqVPD7zc6B+piYmDxfo8Unp9NpTb76Y2NjzZAhQzzeB4pPLpfLREZGBlxHadRddB3F+1q0aGFycnJMXFyckWSio6ONw+EI+NiLTjExMSYlJcX07t3b1KlTx9SsWdNUrFjRxMbGmuTkZOs9d/v27aZBgwYmNjbWSDIpKSnmnnvuMfn5+cYYY2rVqmX1JSYmmoEDB5pdu3Z5va+3adPGJCUlGUlm3bp1VnvxusaNG2fNV7VqVRMfH28iIiKsWouKj4831apVM+3atbPqdn+G9O3b19SuXdukp6ebmJgYU69ePZOcnGweeeQRM2fOHONwOExsbKxxuVymevXqZuDAgSYuLs7UrFnTevxHjx41LVu2tPaLmjVr+ty+SUlJpm7duiYyMtIYY8y7775r2rZta2JiYkxERISJioqyluFr27o/B+rVq+exbbOzs021atVMWlqaR51Ft292drbp2bOnVWdJt+1ll11mKlSo4LEfFF1mr169vLbBunXrTHZ2tvV8F1/2FVdc4Xd/+3//7/9Zyx84cGCJ94VPP/30lMssWm/xmnv27Gl69erlc/nNmzc3DRs29Lv8u+++2yQnJ3ttA2OMSUlJOeN6gynoI28FBQUqLCz0at+3b582btyoJ554Qlu2bNHixYvVt29f7du3T/v27dOmTZv097//XZs3b9bVV1+tNm3a6MiRI8rPz9frr7+uhQsX6oknntCIESMknfiS3IMHD3r0PfLII9b6jDF6/vnntXDhQklSZGSknE6nNe/evXut3yqNjo5WbGysR71RUVHWNzG7v5BXkiIiItS4cWP9/rv3b0nm5+cH/BLAmJgYj2WdLnf9JW2XTtRbVvw9lrN5jOVJoMfhcDgC9gd6Ts4FxV9Pp+LeXsYYv9vW6XT6fH9x3//IkSOaM2eOjh075vFaj4z846L733//XS1atFBUVJTXul0ul9q0aeOxXDeXy6XKlStbtytWrChJVpsxRhUrVrRebw6HQy1btlT9+vWtdVWvXl0Oh0PNmjXThAkTVKtWLUnSwoULFRUVZS2jS5cu2r17tx599FElJiYqLi5OqampSk5O1n333acNGzbo3//+t66++mpr3ampqR7fV3n8+HFt2bJFxhi1bNlSqampuuqqqzRr1izdeeedkqQaNWroggsuUM2aNTV48GBt3bpVffr08druP/zwg88r9ipUqKBWrVqpa9eu+v777zVx4kRJ0oIFC7R3715VrlxZnTt31ieffKLrrrvOa/7CwkL17dtXf/nLXzza9+/fL2OMZsyYobVr1+rxxx/XgQMH9N5771nP/QUXXKAvv/xSCxYs0HfffaejR4/KGGP9buUdd9zh9U36xhhNmjRJu3fvtib3CIwxRm+88YZuuOEGDR48WBs2bFD16tUVFRUlh8Ph9V2g7m1bq1YtJSQk6PHHH9esWbN01113WfepUqWK5s2bp82bN+uNN97Qd999p379+nksZ9OmTapevbrXtnFv1x07dmj37t0e2/bzzz9X7dq1tXbtWn322Wde29YYo4KCAo0dO9bnNrj33nuVm5trbYOJEyeqatWquuKKK/Tjjz8qNzdXP/zwg4YOHar09HRlZmZa869Zs8baF44f/+M3k6dMmaKrr75aXbt21a5duzRu3Di1b9/eWke/fv3UtWtXDRs2zGuZxhjreS1es/s5nTJlinbu3Gktr0WLFoqPj1diYqJ2795t9bmXX7duXb/7gXTi9Vx0P5g4caJVb9H1+KrXXZevDHA2fb7uXGKdOnUyI0aMMCNGjDDJycmmUqVKZsKECaawsNBjZMo9TZgwwTz//PM+/1qtVKmS6dGjh0lKSrL+wvQ1b/F26cQI1ejRo/32+WqXZNLT0/32FR1FK+lfuExM58LkaxS5+OTrNewelfA1hdtrzNf7X6imhISEkK7f1xEJ9xRoBCfQ/hJoOtN9qeiRleL1x8TE+J2vXbt2fvvat2/vt++SSy7x+xivuOIK07lz5zLZ5vHx8X77/vSnPwVc7oIFC0KyDwV6XgPtJ//85z/99t15551G8hz1joiIMNOmTbNuDxw40Pp/enq6ef311829995rWrZsaR566CFrZDIuLs40adLEVKhQwUyZMsW88sorRjrx2nM/Fw888IAxxgTsc4/mLV682LRp08ZERUWZTz75pER57LTDW0JCghk1apTZsmWLefnll01cXJyZPXu26dChg3E4HOaiiy4yn3/+uXnqqadMXFycufHGG01kZKRxOBwmKirK9O/f3yxdutS8/PLLJjo62jidTpOYmGikE2+Ab731lse87g2ZlJRknn32WTN8+HDjcDjMoEGDrL7o6Ghz4YUXmqFDh3o94U6n01x44YUeh0eLTjVr1vTaUdw7h3tD+9rBIyIiAr6IAx2aOd2ppIdqzvSQDtO5MZ3ph9yZzhfoA8XX4ZnTmXyFxWBPkZGRAR+jr8M57ulUQaUsw22gdUdHR/s8bOye50zrio6ONtHR0SYjI8Orz+l0+ny/dK/LVz3u9zp/73nVqlXzW2tsbKzf+dztRectft+iz3nx7ZKamupRf6A/fErzOW7durVX26kO//tbf9H2ihUrnnKZgV4D/h5/9erVvdbvfk/w9dxUrlz5lNvTV00Oh8O0b9/eREREmEaNGlnL79ixo5FkOnXq5LG+qKgoqw53gG/btq2Ji4szVapUMRdddJGZP3++2bp1q7nzzjtNRESE2blzp7nooouMJNOwYUMzd+5cM3v2bPP2228bY0zAPnd4a9GihVmyZInZunWr2bt3b9mEtyZNmpjCwkKrbdy4caZJkyamU6dOJioqykybNs2jz70zR0dHm2rVqnnM261bNyPJtGnTxjidTpOUlGSioqJMZmamyczMtI49F+2rVKmSqVq1qsdx6c6dO5tatWqZiIgIrzf15ORkU6tWLeN0Oj12Mvf/3SN1TqfTmrf4Me/o6Gifb3j+dtpAf7kVPU8i0BTohR3oxRIOU7iNyjAFZ/L1wVsaU6BRGPcfnr6mFStW+O2rUqXKGa2vJP3+tos7IBX9oC/6ISed+HCuXr26z+X4C8wtW7b0O9rXqFEjq94GDRp4LMPhcJhLLrnEdO/e3efz1rhxYzNs2DCfy73yyiv99kVFRZkGDRp4vU86HA7jdDrNfffdZ8aNG+fR7t5Oy5YtM4MHD/b4QE9PT/faTkW3X2RkpBUGij/21q1b+z2q07lzZ1O1alWfj6Fx48Z+n+fY2Fiv/c7X+gcMGGCSk5M91ul+nu666y4zdOhQjz+iUlNTTfPmzc2yZcvMeeed57Es9+dkZGSkVbP7cXXq1MlrXyv6GH09Bl/1xsbG+t3HOnbsaEaOHOmzb8qUKWbu3LlG+mOU0b2vJyUlGWOMufDCC63gN2jQIJORkWFcLpcxxphWrVpZy8rOzjbXXnutiY+PNxs2bLDOd3Q4HKZJkyZm0qRJVn7JysoyWVlZxhhj6tWrZySZRYsWeeWmQH3u8Oar71ROO7wNHjzYo23RokUmMjLSXHLJJSY+Pt7jooRFixYZp9NpoqKijMvlMm3btjU7duww8fHxJj4+3krx7nknTJhgYmNjTXR0tMcOHx8fbx599FEzZ84crxdAdHS0SU9PN2vXrvXaedzL+OSTT3y+6CSZpk2beu0MpwpY7rrPP/98v/e54IIL/PYFGlo/mxEF9yiir+lMD0kwld0U6sNbdpxO9YfLqYLb6fzhc67+EVH8g1s68Z7nb9QjIiLirI40BHpO/D0HdevW9RoVOtspKirK5/rs/DoNp33Y137icDhMTEyM9TiLPl73YE7Tpk09ckFiYqJJTk722Td37lwzZ84cExERYWUU937/+OOPm8jISBMbG2tlgFatWhljjBk7dqyRTvxh5R69u/LKKz36MjMzzT333GM2bNjgFd5++OGH0w5vQblgoejJyampqVq/fr3Wr1+vAQMGeNyvYsWK2rhxo7766iv94x//8Dgh2Ol06sCBA17LPnbsmL7//nvNnz9fkjxOVnb//4477lB8fLzHfO6+zZs3ey2zsLDQ42RLyfPkZPcJhVu2bPH7mFevXu23b8WKFX77Dh8+7LfvVHbt2uW3r/jjQegdPHgw1CWUS4Eu9PB1sVNR5hQ/GHOq+X3V4OtrA5xOp6KiojwubCiq+PtN0eWG+kKe2rVr+11/69at1bp1a6+++vXr66qrrvI5X2FhofLz872W5XA4lJaWpri4OJ/zRUdHq0aNGlZ/8ceflJTk92eNCgsLPd4r3fO6l+V0OlWjRg2PvlNdnHX8+HF17NhRkydP9qiloKDA+n96err1/+TkZOtilujoaK/l1a1b12s/iImJsWpKSEjwWE9iYmLA+txef/11VapUybrtXoZ7Xyy6zRISEqx1SoG3QXR0tPU4UlJSPPqSkpK85nd/rgeqOzU11e/zHxUV5VVPZGSkIiIiFBER4bU/+HrtGmMUGRmpjh07Kj4+Xm3btvXoP3z4sOrUqaOMjAxrXSkpKdq/f78OHz6szZs3eyx39uzZkk68ftevX6+BAwda26Jdu3batm2bJkyYYP3eea9evSRJ3bt3lySNGTNGnTt31vHjx9W+fXuPvgEDBmjTpk3KzMzUk08+6VGnv/eLQE47vK1atcrrdsOGDeVwOHT8+HGPHX3VqlWqVq2aHA6HIiIi9P333ysyMlINGjRQgwYNtG/fPutJcocLd9+OHTusN81jx46poKBAL7zwgsaMGaMhQ4ZYO1NkZKTq16+vZ5991uqT/tihq1Wrpi+++MLnE5uQkGD9v1evXmrXrp1Hv/tF0KBBA2u5Rfvcbx5Ff7je4XB4XElYfOcM9OIpurPGxcX5vW9kZKTHFXLFBVrn6XxA+HpDKmt2vhK1LK/cPR1leXVxWT1G9/PudDpVsWJFv/tBcnKy0tLSfO7/s2bNkiS1aNHC57IleXyQudfnjzsI/u9///Pqi4+PV2RkpLKzs/3O26lTJ599ERERuv766/2urzS2cdFQWfx17H5fc7lcHusyxig+Pt4rMDmdTn399dfasGGDR3vR7err6mFz8irXos9V0ffcgoICtWnTRo8++qh1/6LLrlmzpi699FKfj69hw4bWc+f+sHfPFxUVpczMTO3evdvql068byYkJGjTpk264oorvB5HZGSkli9frl9//dXjw7RKlSrWMvbt2+f1GCMiInTLLbd41RgZGanjx4/7DIIOh0OdO3fWtdde69UnSf/+97+Vlpbm87F36dLF4zPHvd3c29n9b8OGDXXgwAEdPXrUuq/7c/NPf/qT9djcatWqpa5du+rqq6/2epzufaXo8+d+LTVv3tzvZ0V6errH45L+eM05nU6P4JSUlKRq1aqpsLDQ48pg9/aYNGmSJM/9LjIyUvPnz9fy5ct16NAhHTp0yOqLiopSXFycPvjgA23evNl6rBdffLHuu+8+q/6GDRta89SrV0/SiX2mQYMG2rJli7Uv1KhRQ2lpabrrrrtUWFiotm3b6s033/R4bGPHjtWSJUs0duxYvfHGGx59Q4YM0ZtvvqkxY8bomWee8bm9TsvpDNO5L1gYPXq02bJli5k3b56Jj483s2bNsk78q1evnvnss8/M008/beLj482gQYNMcnKyqVixonE6neb//u//zMqVK828efOMy+UyMTEx1jHzypUrm9mzZ5v777/fxMTEWN8XI/1xzP1vf/ubR1/R73UaNmyY12HHosOeRc8hcM/jPievd+/e1lBo8UMAV155ZcDzTIoekgynYWomJjtNLpfrrC+EKOl0qte5w+HwOHm9+FRez1uNiorye87bmUzFT3B3n7ztngJ9L9/pvJcWf8/2957sdDrNsGHD/J6HFh0dbWJiYjyWV/yiBF/nObo/dwKdy1h8Kn4edtGpXbt2fh9/5cqVvbarr5oCfWbFxcV5Ld/hcJjIyEiPbVe8vkqVKnktK9C+HOiiieKH4Ys/h0Xr69evn99vkqhataqJiIjw2Pbu70V0Op3WdnCf61mhQgXTp08fExkZaRISEjzWU6FCBesChVtvvdU4HA4TFxdnkpKSzKhRo8zixYvNSy+9ZCSZjIwMc/XVVxtjjOnbt6+RZNavX2/WrFlj2rZtW6I+92HT33777bQPm552eLv55pvN8OHDTVJSkqlYsaK58847TWFhoenUqZPp06ePx5N75513mueff94kJyeb1q1bm5SUFOuJrlixounWrZtJSkoynTp1Ms2aNfN4oitUqGAuu+wy43Q6TbNmzbw2sLuvT58+Hi/+ov/v0aOHx5t50b569ep5BT33k1y7dm2vE0mL7jjFd/qiywl07kegryrx9aLwtz4mJqYzm073ZP/yOgV6HKF+vwj0YZ6VleW3r6zOy/UVDiMiIqyvCjndc+7KIngnJCT4Dc2BAlBJagr0B02bNm0CLrdt27Z++3z9cVKa37Lg6zEFulJY+uMre4rPW6tWLdOjRw8jyeOc+ho1apgnnnjCut+4ceM8tleNGjVMp06dTEZGhrnllltM/fr1rYtUbrjhBuvK0N69e1uPv0qVKiXuC2p4GzVq1Gn3nc28ZdFHPfaqtSyXW9p1ldfHWJ6Vt20mnfjD73T73EcQFi5cWOI+968G+FpuoD53/6lqda+v+OMN1Ofuj4uLK9U+Y4yZM2eO9QFbvO7zzjvPGo0pvp0C9bkfw5n2jRo1ymd/SbftqfqnTp1qatWqVaZ9Re9z3XXXnVFf0cdTdPml0eer3qL7QkpKyhn1+XoOAj03Z9pXHpWPk3QAAAgDM2bMkCT99NNPeumll/TII49Y50aWVd/q1au1bds2SdKiRYtK1Ffct99+q23btnkt/0z6SlLvnj17dOzYMUmyzm0M1AdP/s96BwAAp+Xbb7+VJI0cOVLp6ekaM2aMxo8fX6Z9U6dO1a+//irpxMV37pP7A/UVl5OTo7vvvlu1a9f2WP6Z9N1+++2nrPeXX36xLlhwX7gRqA+eHMac4tp6AAAAlBscNgUAALARwhsAAICNEN4AAABshPAGAABgI4Q3AAAAGyG8AUAZuvHGG/3+qDsAnAnCGwAAgI0Q3gAAAGyE8AYgbMyfP18ZGRmKjY1VSkqKLr/8ch06dEirV69Wly5dVLlyZSUnJ6tTp05au3atx7wOh0NPP/20evToobi4ODVp0kQrV67U1q1b1blzZ8XHx6tdu3b67rvvrHkmTZqkVq1a6emnn1ZaWpri4uLUv39//e9///NbozFGDz/8sOrVq6fY2Fi1bNlS8+fPL6tNAiAMEd4AhIXdu3drwIABGjJkiP773/9q2bJl6tOnj4wxOnDggLKzs/X5559r1apVatiwobp166YDBw54LOO+++7ToEGDtH79ejVu3FjXXXed/vKXv2j8+PH64osvJEm33HKLxzxbt27V66+/rnfeeUeLFy/W+vXrNWLECL91Tpw4UXPmzNHMmTP11VdfafTo0Ro4cKA+++yz0t8oAMJTGf/wPQAExZo1a4wk8/3335/yvr///rtJTEw077zzjtUmyUycONG6vXLlSiPJPPfcc1bbK6+8YmJiYqzb9957r3E6nSY3N9dq++CDD0xERITZvXu3McaY7Oxs06tXL2OMMQcPHjQxMTFmxYoVHvUMHTrUDBgw4PQeMIBzFiNvAMJCy5YtddlllykjI0P9+/fXM888o99++02StGfPHg0fPlyNGjVScnKykpOTdfDgQe3cudNjGS1atLD+X61aNUlSRkaGR9vRo0eVl5dntdWuXVu1atWybrdr106FhYX6+uuvvWrcvHmzjh49qi5duighIcGaXnzxRY/DsQAQSGSoCwCA0uB0OrV06VKtWLFCS5Ys0ZNPPqkJEyboP//5j0aMGKFffvlF06dPV506deRyudSuXTsdO3bMYxlRUVHW/x0Oh9+2wsJCv3W47+P+tyj3fO+9955q1qzp0edyuU7n4QI4hxHeAIQNh8OhDh06qEOHDrrnnntUp04dLVy4UJ9//rlmzJihbt26SZJyc3O1d+/eUlnnzp079eOPPyo1NVWStHLlSkVERKhRo0Ze923atKlcLpd27typTp06lcr6AZx7CG8AwsJ//vMfffzxx+ratauqVq2q//znP/rll1/UpEkTNWjQQC+99JIyMzOVl5en22+/XbGxsaWy3piYGGVnZ+vRRx9VXl6eRo4cqauvvlrVq1f3um9iYqLGjh2r0aNHq7CwUBdffLHy8vK0YsUKJSQkKDs7u1RqAhDeCG8AwkJSUpKWL1+u6dOnKy8vT3Xq1NFjjz2mrKwsVa9eXTfddJNat26t2rVr64EHHtDYsWNLZb0NGjRQnz591K1bN/3666/q1q2bZsyY4ff+9913n6pWraqcnBxt27ZNFSpU0Pnnn6+77rqrVOoBEP4cxhgT6iIAwI4mTZqkRYsWaf369aEuBcA5hKtNAQAAbITwBgAAYCMcNgUAALARRt4AAABshPAGAABgI4Q3AAAAGyG8AQAA2AjhDQAAwEYIbwAAADZCeAMAALARwhsAAICNEN4AAABs5P8DgQ0WQfHxQ8UAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.rcdefaults()\n",
    "\n",
    "sns.barplot(summary_all, x = \"sample\", y = \"numreads\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set rcParams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# X, Hue, Row\n",
    "X = \"#rname\"\n",
    "HUE = \"tissue\"\n",
    "ROW = \"source\"\n",
    "\n",
    "plt.rcdefaults()\n",
    "rcParams[\"figure.dpi\"] = 300\n",
    "figwidth = len(set(result[X])) * len(set(result[HUE]))\n",
    "rcParams[\"figure.figsize\"] = [figwidth, 10]\n",
    "# rcParams[\"figure.figsize\"] = {\"egr\": [40, 10], \"ptr\": [40, 10]}[SPECIES]\n",
    "rcParams[\"font.size\"] = 22\n",
    "group_palette = {\n",
    "    \"source\": {\n",
    "        \"box\": {\"lab1136\": \"brown\", \"SRA\": \"gray\"},\n",
    "        \"strip\": {\"lab1136\": \"indianred\", \"SRA\": \"black\"}\n",
    "    },\n",
    "    \"tissue\": {\n",
    "        \"box\": {\"x\": \"brown\", \"p\": \"tomato\", \"s\": \"greenyellow\", \"l\": \"green\"},\n",
    "        \"strip\": {\"x\": \"brown\", \"p\": \"tomato\", \"s\": \"greenyellow\", \"l\": \"green\"}\n",
    "    }\n",
    "}\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# source_palette = {\n",
    "#     \"box\": {\"lab1136\": \"brown\", \"SRA\": \"gray\"},\n",
    "#     \"strip\": {\"lab1136\": \"indianred\", \"SRA\": \"black\"}\n",
    "# }\n",
    "# tissue_palette = {\n",
    "#     \"x\": \"brown\",\n",
    "#     \"p\": \"tomato\",\n",
    "#     \"s\": \"greenyellow\",\n",
    "#     \"l\": \"green\"\n",
    "# }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Y = \"numreads\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Y\n",
    "Y = \"numreads\"\n",
    "\n",
    "# TYPE OF PLOTS\n",
    "PLOT_BOX = True\n",
    "PLOT_STRIP = True\n",
    "\n",
    "# \n",
    "aspect = rcParams['figure.figsize']\n",
    "ratio = aspect[0] / aspect[1]\n",
    "g = sns.FacetGrid(\n",
    "    result,\n",
    "    row=ROW,\n",
    "    aspect=ratio,\n",
    "    height=rcParams['figure.figsize'][1],\n",
    "    margin_titles=True,\n",
    "    legend_out=False\n",
    ")\n",
    "\n",
    "if PLOT_BOX:\n",
    "    g.map_dataframe(\n",
    "        sns.boxplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"box\"],\n",
    "        hue_order=list(group_palette[HUE][\"box\"].keys()),\n",
    "        fliersize=0,\n",
    "    )\n",
    "    \n",
    "if PLOT_STRIP:\n",
    "    g.map_dataframe(\n",
    "        sns.stripplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"strip\"],\n",
    "        hue_order=list(group_palette[HUE][\"strip\"].keys()),\n",
    "        size=10,\n",
    "        dodge=True\n",
    "    )\n",
    "\n",
    "# add legend\n",
    "g.add_legend()\n",
    "\n",
    "# save figure\n",
    "plt.savefig(f\"pyplt/{SPECIES.capitalize()}_y={Y}_x={X}_hue={HUE}_row={ROW}.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Y = \"covbases\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Y\n",
    "Y = \"covbases\"\n",
    "\n",
    "# TYPE OF PLOTS\n",
    "PLOT_BOX = True\n",
    "PLOT_STRIP = True\n",
    "\n",
    "# \n",
    "aspect = rcParams['figure.figsize']\n",
    "ratio = aspect[0] / aspect[1]\n",
    "g = sns.FacetGrid(\n",
    "    result,\n",
    "    row=ROW,\n",
    "    aspect=ratio,\n",
    "    height=rcParams['figure.figsize'][1],\n",
    "    margin_titles=True,\n",
    "    legend_out=False\n",
    ")\n",
    "\n",
    "if PLOT_BOX:\n",
    "    g.map_dataframe(\n",
    "        sns.boxplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"box\"],\n",
    "        hue_order=list(group_palette[HUE][\"box\"].keys()),\n",
    "        fliersize=0,\n",
    "    )\n",
    "    \n",
    "if PLOT_STRIP:\n",
    "    g.map_dataframe(\n",
    "        sns.stripplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"strip\"],\n",
    "        hue_order=list(group_palette[HUE][\"strip\"].keys()),\n",
    "        size=10,\n",
    "        dodge=True\n",
    "    )\n",
    "\n",
    "# add legend\n",
    "g.add_legend()\n",
    "\n",
    "# save figure\n",
    "plt.savefig(f\"pyplt/{SPECIES.capitalize()}_y={Y}_x={X}_hue={HUE}_row={ROW}.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Y = \"meandepth\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Y\n",
    "Y = \"meandepth\"\n",
    "\n",
    "# TYPE OF PLOTS\n",
    "PLOT_BOX = True\n",
    "PLOT_STRIP = True\n",
    "\n",
    "# \n",
    "aspect = rcParams['figure.figsize']\n",
    "ratio = aspect[0] / aspect[1]\n",
    "g = sns.FacetGrid(\n",
    "    result,\n",
    "    row=ROW,\n",
    "    aspect=ratio,\n",
    "    height=rcParams['figure.figsize'][1],\n",
    "    margin_titles=True,\n",
    "    legend_out=False\n",
    ")\n",
    "\n",
    "if PLOT_BOX:\n",
    "    g.map_dataframe(\n",
    "        sns.boxplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"box\"],\n",
    "        hue_order=list(group_palette[HUE][\"box\"].keys()),\n",
    "        fliersize=0,\n",
    "    )\n",
    "    \n",
    "if PLOT_STRIP:\n",
    "    g.map_dataframe(\n",
    "        sns.stripplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"strip\"],\n",
    "        hue_order=list(group_palette[HUE][\"strip\"].keys()),\n",
    "        size=10,\n",
    "        dodge=True\n",
    "    )\n",
    "\n",
    "# add legend\n",
    "g.add_legend()\n",
    "\n",
    "# save figure\n",
    "plt.savefig(f\"pyplt/{SPECIES.capitalize()}_y={Y}_x={X}_hue={HUE}_row={ROW}.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Y = \"meandepth_nonzero\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Y\n",
    "Y = \"meandepth_nonzero\"\n",
    "\n",
    "# TYPE OF PLOTS\n",
    "PLOT_BOX = True\n",
    "PLOT_STRIP = True\n",
    "\n",
    "# \n",
    "aspect = rcParams['figure.figsize']\n",
    "ratio = aspect[0] / aspect[1]\n",
    "g = sns.FacetGrid(\n",
    "    result,\n",
    "    row=ROW,\n",
    "    aspect=ratio,\n",
    "    height=rcParams['figure.figsize'][1],\n",
    "    margin_titles=True,\n",
    "    legend_out=False\n",
    ")\n",
    "\n",
    "if PLOT_BOX:\n",
    "    g.map_dataframe(\n",
    "        sns.boxplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"box\"],\n",
    "        hue_order=list(group_palette[HUE][\"box\"].keys()),\n",
    "        fliersize=0,\n",
    "    )\n",
    "    \n",
    "if PLOT_STRIP:\n",
    "    g.map_dataframe(\n",
    "        sns.stripplot,\n",
    "        x=X,\n",
    "        y=Y,\n",
    "        hue=HUE,\n",
    "        palette=group_palette[HUE][\"strip\"],\n",
    "        hue_order=list(group_palette[HUE][\"strip\"].keys()),\n",
    "        size=10,\n",
    "        dodge=True\n",
    "    )\n",
    "\n",
    "# add legend\n",
    "g.add_legend()\n",
    "\n",
    "# save figure\n",
    "plt.savefig(f\"pyplt/{SPECIES.capitalize()}_y={Y}_x={X}_hue={HUE}_row={ROW}.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [STEP1] Removing ***TRASHY*** lines in the final output from REDItools2.0 using `scripts/remove_blines.py`  \n",
    "There are redundant and meaningless lines in REDItools2.0 final output table"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [STEP2] Filter REDItable using the AWK program (`scripts/redit_filter.sh`)\n",
    "REDItable headers:  \n",
    "1. Region\n",
    "2. Postition\n",
    "3. Reference \n",
    "4. Strand \n",
    "5. Coverage-q25\n",
    "6. MeanQ \n",
    "7. BaseCount[A,C,G,T]\n",
    "8. AllSubs\n",
    "9. Frequency\n",
    "10. gCoverage-q25\n",
    "11. gMeanQ\n",
    "12. gBaseCount[A,C,G,T] \n",
    "13. gAllSubs\n",
    "14. gFrequency  \n",
    "\n",
    "Applied filters:\n",
    "1. Coverage-q25 > 10\n",
    "2. Frequency > 0.1\n",
    "3. gFrequency == 0.00\n",
    "\n",
    "The AWK command:  \n",
    "`awk -F \"\\t\" -v OFS=\"\\t\" '$5 > 10 && $9 > 0.1 && $14 == 0.00 {print}' [Input] > [Output]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    *glob.glob(str(FILTERED_PATH[\"Ptr\"] / \"*.txt\")),\n",
    "    \"\\nand\",\n",
    "    *glob.glob(str(FILTERED_PATH[\"Egr\"] / \"*.txt\")),\n",
    "    sep=\", \"\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualize the characters of the filtered data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the species to process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "Configs"
    ]
   },
   "outputs": [],
   "source": [
    "SPECIES = \"Ptr\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read filtered tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict = {\n",
    "    sample: pd.read_table(\n",
    "        FILTERED_PATH[SPECIES] / sample,\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        names=REDIT_HEADER\n",
    "    ) for sample in SAMPLES[SPECIES]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict[list(df_dict.keys())[-1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualizing the number of editing sites per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcdefaults()\n",
    "rcParams[\"figure.dpi\"] = 300\n",
    "rcParams[\"figure.figsize\"] = [len(df_dict) * .3, 6]\n",
    "rcParams[\"font.size\"] = 12\n",
    "\n",
    "tmp = {\"sample\": [], \"numedsites\": []}\n",
    "for sample, df in df_dict.items():\n",
    "    tmp[\"sample\"].append(sample)\n",
    "    tmp[\"numedsites\"].append(df.shape[0])\n",
    "    \n",
    "ax = sns.barplot(\n",
    "    data=pd.DataFrame(tmp),\n",
    "    x=\"sample\",\n",
    "    y=\"numedsites\",\n",
    "    linewidth=2,\n",
    "    edgecolor=\".5\",\n",
    "    facecolor=(0, 0, 0, 0)\n",
    ")\n",
    "ax.set_title(f\"numedsites ({SPECIES}; Freq > .1; Depth > 10; gFreq == 0)\")\n",
    "\n",
    "# Rotate the tick labels and set their alignment.\n",
    "plt.setp(\n",
    "    ax.get_xticklabels(),\n",
    "    rotation=45,\n",
    "    ha=\"right\",\n",
    "    rotation_mode=\"anchor\"\n",
    ")\n",
    "# color xtick labels by tissue type\n",
    "palette = {\n",
    "    \"x\": \"brown\",\n",
    "    \"p\": \"tomato\",\n",
    "    \"s\": \"greenyellow\",\n",
    "    \"l\": \"green\"\n",
    "}\n",
    "[t.set_color(palette[t._text[-6]]) for t in ax.xaxis.get_ticklabels()]\n",
    "\n",
    "plt.savefig(f\"pyplt/{SPECIES}_y=numedsites_x=sample.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Record locations as sets of tuple `Tuple[Region, Position]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "locs = {sample: set([(t.Region, t.Position) for t in df.itertuples()]) for sample, df in df_dict.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the overlapping percentage between samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sample = len(SAMPLES[SPECIES])\n",
    "mat = pd.DataFrame(\n",
    "    [[0] * n_sample] * n_sample,\n",
    "    index=SAMPLES[SPECIES],\n",
    "    columns=SAMPLES[SPECIES]\n",
    ")\n",
    "for i in SAMPLES[SPECIES]:\n",
    "    for j in SAMPLES[SPECIES]:\n",
    "        num_union = len(locs[i].union(locs[j]))\n",
    "        num_itersection = len(locs[i].intersection(locs[j]))\n",
    "        mat.loc[i, j] = num_itersection / num_union"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualizing the overlapping percentage as heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rcParams[\"figure.figsize\"] = [20, 20]\n",
    "rcParams[\"figure.dpi\"] = 300\n",
    "rcParams[\"axes.titlesize\"] = 20\n",
    "rcParams[\"xtick.labelsize\"] = 14\n",
    "rcParams[\"ytick.labelsize\"] = 14\n",
    "rcParams[\"font.size\"] = 12\n",
    "\n",
    "\n",
    "ax = sns.heatmap(mat[mat != 1], annot=True, fmt=\".2f\")\n",
    "ax.set_title(\"Overlapping percentage between samples\")\n",
    "plt.savefig(f\"pyplt/{SPECIES}_sample_percent_overlap.png\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [STEP3] Identify DE-events (**D**ifferentially **E**dited) between xylem and other samples using the method described in:  \n",
    "\"C. Lo Giudice, M. A. Tangaro, G. Pesole, E. Picardi, Investigating RNA editing in deep transcriptome datasets with REDItools and REDIportal. Nat Protoc 15, 1098-1131 (2020).\"\n",
    "and the [GitHub page](https://github.com/BioinfoUNIBA/QEdit/tree/master/scripts)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_**Samples downloaded from SRA are excluded in following analyses**_"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Static configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_FILTERED = Path(\"shout/filtered_c10_f10_g0/\")\n",
    "PATH_ARCHIVED = Path(\"pyout/archive-20230104/\")\n",
    "PATH_PYFILTED = Path(\"pyout/filtered_c10_f10_g0/\")\n",
    "PATH_LOCUNION = {\n",
    "    \"Ptr\": Path(\"pyout/LOC_UNION_PTR.txt\"),\n",
    "    \"Egr\": Path(\"pyout/LOC_UNION_EGR.txt\")\n",
    "}\n",
    "REDIT_HEADER = [\n",
    "    \"Region\", \"Position\", \"Reference\", \"Strand\", \n",
    "    \"Coverage_q25\", \"MeanQ\", \"BaseCount\", \"AllSubs\", \"Frequency\",\n",
    "    \"gCoverage_q25\", \"gMeanQ\", \"gBaseCount\", \"gAllSubs\", \"gFrequency\"\n",
    "]\n",
    "SAMPLES = {\n",
    "    \"Ptr\": {\n",
    "        \"xylem\": {\n",
    "            \"ptr-tsp1-xr.txt\", \"ptr-tsp2-xr.txt\", \"ptr-tsp3-xr.txt\", \"ptr-tsp4-xr.txt\", \"ptr-tsp5-xr.txt\", \"ptr-tsp6-xr.txt\"\n",
    "        }, \n",
    "        \"phloem\": {\n",
    "            \"ptr-tsp1-pr.txt\", \"ptr-tsp2-pr.txt\", \"ptr-tsp3-pr.txt\", \"ptr-tsp4-pr.txt\", \"ptr-tsp5-pr.txt\", \"ptr-tsp6-pr.txt\"\n",
    "        }, \n",
    "        \"leaf\": {\n",
    "            \"ptr-tsp1-lr.txt\", \"ptr-tsp2-lr.txt\", \"ptr-tsp3-lr.txt\", \"ptr-tsp4-lr.txt\", \"ptr-tsp5-lr.txt\", \"ptr-tsp6-lr.txt\"\n",
    "        }, \n",
    "        \"shoot\": {\n",
    "            \"ptr-tsp1-sr.txt\", \"ptr-tsp2-sr.txt\", \"ptr-tsp3-sr.txt\", \"ptr-tsp4-sr.txt\", \"ptr-tsp5-sr.txt\", \"ptr-tsp6-sr.txt\"\n",
    "        }\n",
    "    },\n",
    "    \"Egr\": {\n",
    "        \"xylem\": {\n",
    "            \"egr-tsp4-xr.txt\", \"egr-tsp5-xr.txt\", \"egr-tsp7-xr.txt\"\n",
    "        },\n",
    "        \"phloem\": {\n",
    "            \"egr-tsp4-pr.txt\", \"egr-tsp5-pr.txt\", \"egr-tsp7-pr.txt\"\n",
    "        },\n",
    "        \"leaf\": {\n",
    "            \"egr-tsp4-lr.txt\", \"egr-tsp5-lr.txt\", \"egr-tsp7-lr.txt\"\n",
    "        },\n",
    "        \"shoot\": {\n",
    "            \"egr-tsp4-sr.txt\", \"egr-tsp5-sr.txt\", \"egr-tsp7-sr.txt\"\n",
    "        }        \n",
    "    }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Params per analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Species to process\n",
    "SPECIES = \"Ptr\"\n",
    "\n",
    "# Output DEtable\n",
    "FOUT = Path()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the union of locations of all samples"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "d3bfaef271ab48e0cc0541a0a3727f5392cd9dce4b40e4c93919be8e988540f3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
